// @bbva-web-components/bbva-list-multistep/bbva-list-multistep/bbva-list-multistep-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  display: block;
  box-sizing: border-box;
  padding: 0 calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1em) calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1em); }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

.row {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  flex-grow: 1; }

.column {
  display: flex;
  flex-direction: column;
  flex-grow: 1; }

.card > .column,
.contact > .column {
  margin-left: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem); }

.balance {
  text-align: right; }

.name,
.content ::slotted([slot='number']),
.content ::slotted([slot='description']),
.content ::slotted([slot='details']) {
  font-size: var(--typographyTypeSmall, ${unsafeCSS(foundations.typography.typeSmall)});
  line-height: var(--lineHeightTypeSmall, ${unsafeCSS(foundations.lineHeight.typeSmall)}); }

.content ::slotted([slot='description']) {
  font-style: var(--fontFacePrimaryBookFontStyle, ${unsafeCSS(foundations.fontFacePrimary.book.fontStyle)});
  font-weight: var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(foundations.fontFacePrimary.book.fontWeight)});
  color: var(--bbva-list-multistep-description-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

.card .wrapper-name {
  min-width: 0;
  flex-basis: 35%;
  margin-right: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem); }

.name {
  font-style: var(--fontFacePrimaryMediumFontStyle, ${unsafeCSS(foundations.fontFacePrimary.medium.fontStyle)});
  font-weight: var(--fontFacePrimaryMediumFontWeight, ${unsafeCSS(foundations.fontFacePrimary.medium.fontWeight)});
  display: block;
  text-overflow: ellipsis;
  overflow: hidden;
  white-space: nowrap;
  color: var(--bbva-list-multistep-name-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

.content ::slotted([slot='number']),
.content ::slotted([slot='details']) {
  font-style: var(--fontFacePrimaryBookItalicFontStyle, ${unsafeCSS(foundations.fontFacePrimary.bookItalic.fontStyle)});
  font-weight: var(--fontFacePrimaryBookItalicFontWeight, ${unsafeCSS(foundations.fontFacePrimary.bookItalic.fontWeight)}); }

.content ::slotted([slot='details']) {
  color: var(--bbva-list-multistep-slotted-details-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

.amount {
  color: var(--bbva-list-multistep-amount-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

.content ::slotted([slot='number']) {
  display: flex;
  color: var(--bbva-list-multistep-slotted-number-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([variant='maximum']) .row:not(.row-small) {
  flex-direction: column; }

:host([variant='maximum']) .description {
  order: 3; }

:host([variant='maximum']) .content ::slotted([slot='number']) {
  order: 4; }

:host([variant='maximum']) .card .clip {
  margin: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem) 0; }
`;
