// @bbva-web-components/bbva-form-select/bbva-form-select/bbva-form-select.js
/* eslint-disable max-classes-per-file */
import { html } from 'lit-element';
import { getComponentSharedStyles } from '@cells-components/cells-lit-helpers';
import { CellsSelect, CellsOption } from '@cells-components/cells-select';
import '@cells-components/cells-icon';
import { bbvaFoundationsStylesFocus } from '@bbva-web-components/bbva-foundations-styles';
import styles from './bbva-form-select-styles.js';
import stylesOption from './bbva-form-option-styles.js';

export { bbvaFormSelectAmbient } from './bbva-form-select-ambient.js';
/**
Dropdowns are used to let users select an option from a set of choices. They should be primarily used if there are 3 or more options. For anything with fewer options consider using radio buttons or checkboxes.

`bbva-form-select` extends from `cells-select` to show an element with select dropdown functionality, providing it with BBVA Experience styles and structure. As it extends from `cells-select`, this select has all the basic select behaviors as autofocus, form participation, required, disabled & readonly attributes...

Also, it provides a custom `bbva-form-option` to be used declaratively just as a native `select` with `options`. `bbva-form-option` has a `clip` slot to provide a clip element in each option, which will be shown in the select when the option is selected.

Example:
```html
<bbva-form-select label="Default" name="select1">
  <bbva-form-option value="option1">Option 1</bbva-form-option>
  <bbva-form-option value="option2" selected>Option 2</bbva-form-option>
  <bbva-form-option value="option3">Option 3</bbva-form-option>
</bbva-form-select>
```

## Styling

The following custom properties and mixins are available for styling:

### Custom properties

| Selector                         | CSS Property                    | CSS Variable                               | Theme Variable                  | Foundations/Fallback                                      |
| -------------------------------- | ------------------------------- | ------------------------------------------ | ------------------------------- | --------------------------------------------------------- |
| .info .message-icon              | color                           | --_info-message-icon-color                 |                                 |                                                           |
| .message-icon                    | margin-right                    | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .message-icon                    | color                           | --_message-icon-color                      |                                 |                                                           |
| .message.info                    | color                           | --_info-message-color                      |                                 |                                                           |
| .message                         | padding                         | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .message                         | font-size                       |                                            | --typographyType3XSmall         | foundations.typography.type3XSmall          |
| .message                         | line-height                     |                                            | --lineHeightType3XSmall         | foundations.lineHeight.type3XSmall          |
| .message                         | color                           | --_message-color                           |                                 |                                                           |
| .options-wrapper                 | background-color                |                                            | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .options-gradient::after         | background                      |                                            | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .options-gradient::before        | background                      |                                            | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .options-overlay                 | background-color                |                                            | --colorsPrimaryCoreDarkened     | foundations.colors.primaryCoreDarkened      |
| .content-value                   | margin-left                     | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| [aria-invalid=true] .label       | color                           | --_invalid-select-label-color              |                                 |                                                           |
| :host([readonly]) .label         | color                           | --_select-readonly-label-color             |                                 |                                                           |
| :host([filled]) .label           | color                           | --_select-filled-label-color               |                                 |                                                           |
| :host([filled]) .label           | transform                       | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .label                           | color                           | --_select-label-color                      |                                 |                                                           |
| .label                           | transform                       | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| [aria-invalid=true] .button-icon | color                           | --_invalid-select-icon-color               |                                 |                                                           |
| :host([readonly]) .button-icon   | color                           | --_select-readonly-icon-color              |                                 |                                                           |
| .button-icon                     | margin-right                    | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button-icon                     | color                           | --_select-icon-color                       |                                 |                                                           |
| .button-text                     | padding                         | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button-text                     | margin-left                     | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button-clip                     | margin-left                     | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button[aria-invalid=true]       | background-color                | --_invalid-select-bg-color                 |                                 |                                                           |
| .button[aria-invalid=true]       | border-bottom-color             | --_invalid-select-border-color             |                                 |                                                           |
| :host([readonly]) .button        | color                           | --_select-readonly-button-color            |                                 |                                                           |
| :host([readonly]) .button        | border-bottom-color             | --_select-readonly-border-color            |                                 |                                                           |
| .button                          | font-size                       |                                            | --typographyTypeSmall           | foundations.typography.typeSmall            |
| .button                          | line-height                     |                                            | --lineHeightTypeSmall           | foundations.lineHeight.typeSmall            |
| .button                          | font-weight                     |                                            | --fontFacePrimaryBookFontWeight | foundations.fontFacePrimary.book.fontWeight |
| .button                          | background-color                | --_select-bg-color                         |                                 |                                                           |
| .button                          | color                           | --_select-button-color                     |                                 |                                                           |
| .button                          | border-bottom                   | --_select-border-color                     |                                 |                                                           |
| :host                            | --_select-bg-color              | --bbva-form-select-bg-color                | --colorsSecondary100            | foundations.colors.secondary100             |
| :host                            | --_select-border-color          | --bbva-form-select-border-color            | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-readonly-border-color | --bbva-form-select-readonly-border-color   | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-button-color          | --bbva-form-select-button-color            | --colorsSecondary600            | foundations.colors.secondary600             |
| :host                            | --_select-readonly-button-color | --bbva-form-select-readonly-button-color   | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-label-color           | --bbva-form-select-label-color             | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-filled-label-color    | --bbva-form-select-filled-label-color      | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-readonly-label-color  | --bbva-form-select-readonly-label-color    | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-icon-color            | --bbva-form-select-icon-color              | --colorsPrimaryMedium           | foundations.colors.primaryMedium            |
| :host                            | --_select-readonly-icon-color   | --bbva-form-select-readonly-icon-color     | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_message-color                | --bbva-form-select-message-color           | --colorsSecondary600            | foundations.colors.secondary600             |
| :host                            | --_message-icon-color           | --bbva-form-select-message-icon-color      | --colorsTertiaryType1Dark       | foundations.colors.tertiaryType1Dark        |
| :host                            | --_info-message-color           | --bbva-form-select-info-message-color      | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_info-message-icon-color      | --bbva-form-select-info-message-icon-color | --colorsPrimarySubdued          | foundations.colors.primarySubdued           |
| :host                            | --_invalid-select-bg-color      | --bbva-form-select-invalid-bg-color        | --colorsTertiaryType5Lightened  | foundations.colors.tertiaryType5Lightened   |
| :host                            | --_invalid-select-border-color  | --bbva-form-select-invalid-border-color    | --colorsTertiaryType5Medium     | foundations.colors.tertiaryType5Medium      |
| :host                            | --_invalid-select-label-color   | --bbva-form-select-invalid-label-color     | --colorsTertiaryType1Dark       | foundations.colors.tertiaryType1Dark        |
| :host                            | --_invalid-select-icon-color    | --bbva-form-select-invalid-icon-color      | --colorsTertiaryType1Dark       | foundations.colors.tertiaryType1Dark        |

### @apply

| Selector                         | Mixins                                             |
| -------------------------------- | -------------------------------------------------- |
| .info .message-icon              | --bbva-form-select-info-message-icon               |
| .message-icon                    | --bbva-form-select-message-icon                    |
| .message.info                    | --bbva-form-select-info-message                    |
| .message                         | --bbva-form-select-message                         |
| .options-wrapper                 | --bbva-form-select-options-wrapper                 |
| .options-gradient::after         | --bbva-form-select-options-gradient-bottom-element |
| .options-gradient::before        | --bbva-form-select-options-gradient-top-element    |
| .options-gradient::before        | --bbva-form-select-options-gradient-element        |
| .options-gradient::after         | --bbva-form-select-options-gradient-element        |
| .options-gradient                | --bbva-form-select-options-gradient                |
| .options-overlay                 | --bbva-form-select-options-overlay                 |
| .options-layer.opened            | --bbva-form-select-opened-options-layer            |
| .options-layer                   | --bbva-form-select-options-layer                   |
| .content-value                   | --bbva-form-select-content-value                   |
| .content-text                    | --bbva-form-select-content-text                    |
| .content                         | --bbva-form-select-content                         |
| [aria-invalid=true] .label       | --bbva-form-select-invalid-label                   |
| :host([readonly]) .label         | --bbva-form-select-readonly-label                  |
| :host([filled]) .label           | --bbva-form-select-filled-label                    |
| .label                           | --bbva-form-select-label                           |
| [aria-invalid=true] .button-icon | --bbva-form-select-invalid-button-icon             |
| :host([readonly]) .button-icon   | --bbva-form-select-readonly-button-icon            |
| .button-icon                     | --bbva-form-select-button-icon                     |
| .button-text                     | --bbva-form-select-button-text                     |
| .button-clip                     | --bbva-form-select-button-clip                     |
| .button-content                  | --bbva-form-select-button-content                  |
| .button[aria-invalid=true]       | --bbva-form-select-invalid-button                  |
| .button[disabled]                | --bbva-form-select-disabled-button                 |
| :host([readonly]) .button        | --bbva-form-select-readonly-button                 |
| .button                          | --bbva-form-select-button                          |
| .slot-select ::slotted(select)   | --bbva-form-select-select                          |
| :host                            | --bbva-form-select                                 |

## Ambient

### Custom properties

| Selector          | CSS Property                               | CSS Variable | Theme Variable                 | Foundations/Fallback                                    |
| ----------------- | ------------------------------------------ | ------------ | ------------------------------ | ------------------------------------------------------- |
| [ambient=dark300] | --bbva-form-select-bg-color                |              | --colorsPrimaryCore            | foundations.colors.primaryCore            |
| [ambient=dark150] | --bbva-form-select-bg-color                |              | --colorsPrimaryMedium          | foundations.colors.primaryMedium          |
| [ambient^=dark]   | --bbva-form-select-border-color            |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-readonly-border-color   |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=dark]   | --bbva-form-select-button-color            |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-readonly-button-color   |              | --colorsSecondary300           | foundations.colors.secondary300           |
| [ambient^=dark]   | --bbva-form-select-label-color             |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-filled-label-color      |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=dark]   | --bbva-form-select-readonly-label-color    |              | --colorsSecondary300           | foundations.colors.secondary300           |
| [ambient^=dark]   | --bbva-form-select-icon-color              |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-readonly-icon-color     |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=dark]   | --bbva-form-select-message-color           |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-message-icon-color      |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-info-message-color      |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-info-message-icon-color |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-invalid-bg-color        |              | --colorsTertiaryType5Lightened | foundations.colors.tertiaryType5Lightened |
| [ambient^=dark]   | --bbva-form-select-invalid-border-color    |              | --colorsTertiaryType5Medium    | foundations.colors.tertiaryType5Medium    |
| [ambient^=dark]   | --bbva-form-select-invalid-label-color     |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
| [ambient^=dark]   | --bbva-form-select-invalid-icon-color      |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |

### Custom properties

| Selector           | CSS Property                               | CSS Variable | Theme Variable                 | Foundations/Fallback                                    |
| ------------------ | ------------------------------------------ | ------------ | ------------------------------ | ------------------------------------------------------- |
| [ambient=light100] | --bbva-form-select-bg-color                |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=light]   | --bbva-form-select-bg-color                |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=light]   | --bbva-form-select-border-color            |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-readonly-border-color   |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-button-color            |              | --colorsSecondary600           | foundations.colors.secondary600           |
| [ambient^=light]   | --bbva-form-select-readonly-button-color   |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-label-color             |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-filled-label-color      |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-readonly-label-color    |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-icon-color              |              | --colorsPrimaryMedium          | foundations.colors.primaryMedium          |
| [ambient^=light]   | --bbva-form-select-readonly-icon-color     |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-message-color           |              | --colorsSecondary600           | foundations.colors.secondary600           |
| [ambient^=light]   | --bbva-form-select-message-icon-color      |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
| [ambient^=light]   | --bbva-form-select-info-message-color      |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-info-message-icon-color |              | --colorsPrimarySubdued         | foundations.colors.primarySubdued         |
| [ambient^=light]   | --bbva-form-select-invalid-bg-color        |              | --colorsTertiaryType5Lightened | foundations.colors.tertiaryType5Lightened |
| [ambient^=light]   | --bbva-form-select-invalid-border-color    |              | --colorsTertiaryType5Medium    | foundations.colors.tertiaryType5Medium    |
| [ambient^=light]   | --bbva-form-select-invalid-label-color     |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
| [ambient^=light]   | --bbva-form-select-invalid-icon-color      |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
> Styling documentation generated by Cells CLI

* @customElement bbva-form-select
* @polymer
* @LitElement
* @demo demo/index.html
*/
export class BbvaFormSelect extends CellsSelect {
  static get is() {
    return 'bbva-form-select';
  }

  /**
   * TagName (uppercase) of select option elements
   * @return {String} TagName of options (uppercase)
   */
  static get optionTag() {
    return 'BBVA-FORM-OPTION';
  }

  static get properties() {
    return {
      /**
       * Icon to show when select is closed
       */
      openIcon: {
        type: String,
        attribute: 'open-icon',
      },

      /**
       * Icon to show when select is open
       */
      closeIcon: {
        type: String,
        attribute: 'close-icon',
      },

      /**
       * Size for open/close icons
       */
      iconSize: {
        type: Number,
        attribute: 'icon-size',
      },

      /**
       * If true, value of selected option will be shown in select control
       */
      showValue: {
        type: Boolean,
        attribute: 'show-value',
      },

      /**
       * Icon for error message
       */
      errorIcon: {
        type: String,
        attribute: 'error-icon',
      },

      /**
       * Size for error icon
       */
      errorIconSize: {
        type: Number,
        attribute: 'error-icon-size',
      },

      /**
       * Error message
       */
      errorMessage: {
        type: String,
        attribute: 'error-message',
      },

      /**
       * Icon for info message
       */
      infoIcon: {
        type: String,
        attribute: 'info-icon',
      },

      /**
       * Size for info icon
       */
      infoIconSize: {
        type: Number,
        attribute: 'info-icon-size',
      },

      /**
       * Info message
       */
      infoMessage: {
        type: String,
        attribute: 'info-message',
      },
    };
  }

  constructor() {
    super();
    this.openIcon = 'coronita:unfold';
    this.closeIcon = 'coronita:fold';
    this.iconSize = 24;
    this.errorIcon = 'coronita:alert';
    this.errorIconSize = 16;
    this.errorMessage = '';
    this.infoIcon = 'coronita:info';
    this.infoIconSize = 16;
    this.infoMessage = '';
  }

  updated(changedProps) {
    /* istanbul ignore else */
    if (super.updated) {
      super.updated(changedProps);
    }
    if (changedProps.has('_selectedOption')) {
      if (this._selectedOptionText) {
        this.setAttribute('filled', '');
      } else {
        this.removeAttribute('filled');
      }
    }
  }

  get _errorVisible() {
    return this.errorMessage && this.invalid;
  }

  get _selectedOptionText() {
    return this._selectedOption && this._selectedOption.text;
  }

  get _selectedOptionClip() {
    return this._selectedOption && this._selectedOption.clip;
  }

  get _selectedOptionValue() {
    return this._selectedOption && this._selectedOption.value;
  }

  static get shadyStyles() {
    return `
      ${bbvaFoundationsStylesFocus('.button:focus')}
      ${styles.cssText}
      ${getComponentSharedStyles('bbva-form-select-shared-styles')}
    `;
  }

  render() {
    return html`
      <style>
        ${this.constructor.shadyStyles}
      </style>
      ${this.field} ${this.messageError} ${this.messageInfo}
    `;
  }

  /**
   * HTML for select field block
   * @return {TemplateResult} HTML of select field block
   */
  get field() {
    return html`
      ${this._control} ${this._optionsLayer} ${this._select}
    `;
  }

  /**
   * HTML for control button content
   * @return {TemplateResult} HTML of control button content
   */
  get _controlContent() {
    return html`
      <span class="button-content">
        ${this._clip}
        <span class="button-text">
          ${this._label}
          <span class="content">
            <span class="content-text">
              ${this._selectedOptionText}
            </span>
            ${this.showValue
              ? html`
                  <span class="content-value">
                    ${this._selectedOptionValue}
                  </span>
                `
              : ''}
          </span>
        </span>
        ${this._selectIcon}
      </span>
    `;
  }

  /**
   * HTML for selected option clip, if available
   * @return {TemplateResult} HTML of selected option clip
   */
  get _clip() {
    if (this._selectedOptionClip) {
      const clip = this._selectedOptionClip.cloneNode(true);
      return html`
        <span class="button-clip" aria-hidden="true">
          ${clip}
        </span>
      `;
    }
    return '';
  }

  /**
   * HTML for select label
   * @return {TemplateResult} HTML of select label
   */
  get _label() {
    return html`
      <span id="label" class="label" aria-hidden="true"
        >${this.label}
        ${this.required
          ? html`
              *
            `
          : ''}</span
      >
    `;
  }

  /**
   * HTML for control icon
   * @return {TemplateResult} HTML of control icon
   */
  get _selectIcon() {
    return html`
      <span class="button-icon" aria-hidden="true">
        <cells-icon
          .icon="${this.opened ? this.closeIcon : this.openIcon}"
          .size="${this.iconSize}"
        ></cells-icon>
      </span>
    `;
  }

  /**
   * HTML for options list container
   * @return {TemplateResult} HTML for options list container
   */
  get _optionsLayer() {
    return html`
      <div class="options-layer ${this.opened ? 'opened' : ''}">
        <div class="options-overlay"></div>
        <div class="options-gradient">
          <div class="options-wrapper">
            ${this._optionsList}
          </div>
        </div>
      </div>
    `;
  }

  /**
   * Template for array-provided options
   * @return {TemplateResult} HTML of items options
   */
  /* eslint-disable-next-line class-methods-use-this */
  _optionItemTemplate(item) {
    return html`
      <bbva-form-option
        ?disabled="${item.disabled}"
        ?selected="${item.selected}"
        .value="${item.value}"
        >${item.text}</bbva-form-option
      >
    `;
  }

  /**
   * Layer with error icon and error message
   * @return {TemplateResult} HTML of error message
   */
  get messageError() {
    return this._errorVisible
      ? html`
          <div class="message error">
            <cells-icon
              class="message-icon"
              .icon="${this.errorIcon}"
              .size="${this.errorIconSize}"
              aria-hidden="true"
            ></cells-icon>
            <span>${this.errorMessage}</span>
          </div>
        `
      : '';
  }

  /**
   * Layer with info icon and info message
   * @return {TemplateResult} HTML of info message
   */
  get messageInfo() {
    return this.infoMessage
      ? html`
          <div class="message info">
            <cells-icon
              class="message-icon"
              .icon="${this.infoIcon}"
              .size="${this.infoIconSize}"
              aria-hidden="true"
            ></cells-icon>
            <span>${this.infoMessage}</span>
          </div>
        `
      : '';
  }
}

customElements.define(BbvaFormSelect.is, BbvaFormSelect);

/*
`bbva-form-option` extends from `cells-option` to provide custom option elements aimed to be used as children of `bbva-form-select`. These options can receive text, value... as normal select options. It also provides a 'clip' slot for showing a clip element besides the option, which will be shown in the select when the option is selected.

## Styling

The following custom properties and mixins are available for styling:

### Custom properties

| Selector              | CSS Property     | CSS Variable             | Theme Variable                  | Foundations/Fallback                                      |
| --------------------- | ---------------- | ------------------------ | ------------------------------- | --------------------------------------------------------- |
| :host([tabindex='0']) | background-color |                          | --colorsPrimaryCore             | foundations.colors.primaryCore              |
| :host([tabindex='0']) | color            |                          | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .icon                 | margin-left      | --gridSpacerVariant      | --gridSpacer                    | foundations.grid.spacer                     |
| .clip ::slotted(*)    | margin-right     | --gridSpacerVariant      | --gridSpacer                    | foundations.grid.spacer                     |
| :host                 | padding          | --gridSpacerVariant      | --gridSpacer                    | foundations.grid.spacer                     |
| :host                 | font-size        |                          | --typographyTypeSmall           | foundations.typography.typeSmall            |
| :host                 | line-height      |                          | --lineHeightTypeSmall           | foundations.lineHeight.typeSmall            |
| :host                 | font-weight      |                          | --fontFacePrimaryBookFontWeight | foundations.fontFacePrimary.book.fontWeight |
| :host                 | font-style       |                          | --fontFacePrimaryBookFontStyle  | foundations.fontFacePrimary.book.fontStyle  |
| :host                 | color            | --bbva-form-option-color | --colorsSecondary600            | foundations.colors.secondary600             |

### @apply

| Selector                    | Mixins                          |
| --------------------------- | ------------------------------- |
| :host([aria-disabled=true]) | --bbva-form-option-disabled     |
| :host([tabindex='0'])       | --bbva-form-option-focused      |
| .icon                       | --bbva-form-option-icon         |
| .text                       | --bbva-form-option-text         |
| .clip ::slotted(*)          | --bbva-form-option-clip-slotted |
| .clip                       | --bbva-form-option-clip         |
| :host                       | --bbva-form-option              |

* @customElement bbva-form-option
* @polymer
* @LitElement
* @demo demo/index.html
*/
export class BbvaFormOption extends CellsOption {
  static get is() {
    return 'bbva-form-option';
  }

  static get properties() {
    return {
      /**
       * Icon to show when option is selected
       */
      icon: {
        type: String,
      },

      /**
       * Size for selected option icon
       */
      iconSize: {
        type: Number,
        attribute: 'icon-size',
      },
    };
  }

  constructor() {
    super();
    this.icon = 'coronita:checkmark';
    this.iconSize = 24;
  }

  /**
   * Returns node assigned in Light DOM to clip slot
   * @return { Node } Node assigned in light DOM to clip slot
   */
  get clip() {
    return this.querySelector('[slot=clip]');
  }

  static get shadyStyles() {
    return `
      ${stylesOption.cssText}
      ${getComponentSharedStyles('cells-option-shared-styles')}
    `;
  }

  render() {
    return html`
      <style>
        ${this.constructor.shadyStyles}
      </style>
      <span class="clip">
        <slot name="clip"></slot>
      </span>
      <span class="text">
        <slot></slot>
      </span>
      ${this.selected
        ? html`
            <span class="icon">
              <cells-icon .icon="${this.icon}" .size="${this.iconSize}"></cells-icon>
            </span>
          `
        : ''}
    `;
  }
}

customElements.define(BbvaFormOption.is, BbvaFormOption);
