// @bbva-web-components/bbva-form-select/bbva-form-select/bbva-form-select-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  --_select-bg-color: var(--bbva-form-select-bg-color, var(--colorsSecondary100, ${unsafeCSS(foundations.colors.secondary100)}));
  --_select-border-color: var(--bbva-form-select-border-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_select-readonly-border-color: var(--bbva-form-select-readonly-border-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_select-button-color: var(--bbva-form-select-button-color, var(--colorsSecondary600, ${unsafeCSS(foundations.colors.secondary600)}));
  --_select-readonly-button-color: var(--bbva-form-select-readonly-button-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_select-label-color: var(--bbva-form-select-label-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_select-filled-label-color: var(--bbva-form-select-filled-label-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_select-readonly-label-color: var(--bbva-form-select-readonly-label-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_select-icon-color: var(--bbva-form-select-icon-color, var(--colorsPrimaryMedium, ${unsafeCSS(foundations.colors.primaryMedium)}));
  --_select-readonly-icon-color: var(--bbva-form-select-readonly-icon-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_message-color: var(--bbva-form-select-message-color, var(--colorsSecondary600, ${unsafeCSS(foundations.colors.secondary600)}));
  --_message-icon-color: var(--bbva-form-select-message-icon-color, var(--colorsTertiaryType1Dark, ${unsafeCSS(foundations.colors.tertiaryType1Dark)}));
  --_info-message-color: var(--bbva-form-select-info-message-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_info-message-icon-color: var(--bbva-form-select-info-message-icon-color, var(--colorsPrimarySubdued, ${unsafeCSS(foundations.colors.primarySubdued)}));
  --_invalid-select-bg-color: var(--bbva-form-select-invalid-bg-color, var(--colorsTertiaryType5Lightened, ${unsafeCSS(foundations.colors.tertiaryType5Lightened)}));
  --_invalid-select-border-color: var(--bbva-form-select-invalid-border-color, var(--colorsTertiaryType5Medium, ${unsafeCSS(foundations.colors.tertiaryType5Medium)}));
  --_invalid-select-label-color: var(--bbva-form-select-invalid-label-color, var(--colorsTertiaryType1Dark, ${unsafeCSS(foundations.colors.tertiaryType1Dark)}));
  --_invalid-select-icon-color: var(--bbva-form-select-invalid-icon-color, var(--colorsTertiaryType1Dark, ${unsafeCSS(foundations.colors.tertiaryType1Dark)}));
  display: block;
  box-sizing: border-box;
  position: relative;
  @apply --bbva-form-select; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *::before, *::after {
  box-sizing: inherit; }

.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

.slot-select ::slotted(select) {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  margin: 0;
  opacity: 0;
  z-index: -1;
  pointer-events: none;
  @apply --bbva-form-select-select; }

.button {
  position: relative;
  width: 100%;
  height: 3rem;
  font-family: inherit;
  font-size: var(--typographyTypeSmall, ${unsafeCSS(foundations.typography.typeSmall)});
  line-height: var(--lineHeightTypeSmall, ${unsafeCSS(foundations.lineHeight.typeSmall)});
  font-weight: var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(foundations.fontFacePrimary.book.fontWeight)});
  background-color: var(--_select-bg-color);
  color: var(--_select-button-color);
  border: none;
  border-bottom: 1px solid var(--_select-border-color);
  border-radius: 1px;
  margin: 0;
  outline: none;
  padding: 0;
  cursor: pointer;
  text-align: left;
  @apply --bbva-form-select-button; }
  .button::-moz-focus-inner {
    border: 0; }
  .button[aria-invalid=true] {
    background-color: var(--_invalid-select-bg-color);
    border-bottom-color: var(--_invalid-select-border-color);
    @apply --bbva-form-select-invalid-button; }
  .button[disabled] {
    opacity: .4;
    cursor: default;
    @apply --bbva-form-select-disabled-button; }
  :host([readonly]) .button {
    color: var(--_select-readonly-button-color);
    border-bottom-color: var(--_select-readonly-border-color);
    cursor: default;
    @apply --bbva-form-select-readonly-button; }
  .button-content {
    display: flex;
    align-items: center;
    height: 100%;
    @apply --bbva-form-select-button-content; }
  .button-clip {
    flex: none;
    max-width: 3rem;
    margin-left: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
    display: flex;
    align-items: center;
    @apply --bbva-form-select-button-clip; }
  .button-text {
    flex: auto;
    position: relative;
    height: 100%;
    padding: calc(((2.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 15) * 1em) 0 calc(((0.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 15) * 1em);
    margin-left: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
    min-width: 0;
    @apply --bbva-form-select-button-text; }
  .button-icon {
    flex: none;
    margin-right: calc(((0.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
    width: 3rem;
    display: flex;
    align-items: center;
    justify-content: center;
    color: var(--_select-icon-color);
    @apply --bbva-form-select-button-icon; }
    [aria-invalid=true] .button-icon {
      color: var(--_invalid-select-icon-color);
      @apply --bbva-form-select-invalid-button-icon; }
    :host([readonly]) .button-icon {
      color: var(--_select-readonly-icon-color);
      @apply --bbva-form-select-readonly-button-icon; }

.label {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  color: var(--_select-label-color);
  will-change: transform;
  transform: translateY(calc(((1.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
  transform-origin: top left;
  transition: transform .2s ease-in-out;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  @apply --bbva-form-select-label; }
  [aria-invalid=true] .label {
    color: var(--_invalid-select-label-color);
    @apply --bbva-form-select-invalid-label; }
  :host([filled]) .label {
    width: 125%;
    color: var(--_select-filled-label-color);
    transform: translateY(calc(((0.3 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem)) scale(0.8);
    transition: transform .2s ease-in-out;
    @apply --bbva-form-select-filled-label; }
  :host([readonly]) .label {
    color: var(--_select-readonly-label-color);
    @apply --bbva-form-select-readonly-label; }

.content {
  display: flex;
  @apply --bbva-form-select-content; }
  .content-text {
    flex: auto;
    min-width: 0;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    @apply --bbva-form-select-content-text; }
  .content-value {
    flex: none;
    margin-left: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
    @apply --bbva-form-select-content-value; }

.options-layer {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 1rem;
  opacity: 0;
  visibility: hidden;
  transition: opacity .2s, visibility 0s .2s;
  @apply --bbva-form-select-options-layer; }
  .options-layer.opened {
    opacity: 1;
    visibility: visible;
    transition: opacity .2s, visibility 0s 0s;
    @apply --bbva-form-select-opened-options-layer; }

.options-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  opacity: 0.95;
  background-color: var(--colorsPrimaryCoreDarkened, ${unsafeCSS(foundations.colors.primaryCoreDarkened)});
  @apply --bbva-form-select-options-overlay; }

.options-gradient {
  position: relative;
  max-width: 18rem;
  width: 100%;
  @apply --bbva-form-select-options-gradient; }
  .options-gradient::before, .options-gradient::after {
    content: '';
    position: absolute;
    left: 0;
    right: 0;
    height: 2.5rem;
    @apply --bbva-form-select-options-gradient-element; }
  .options-gradient::before {
    top: 0;
    background: linear-gradient(to bottom, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}), rgba(255, 255, 255, 0));
    @apply --bbva-form-select-options-gradient-top-element; }
  .options-gradient::after {
    bottom: 0;
    background: linear-gradient(to top, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}), rgba(255, 255, 255, 0));
    @apply --bbva-form-select-options-gradient-bottom-element; }

.options-wrapper {
  overflow-y: auto;
  min-height: 14rem;
  max-height: 29rem;
  padding: 2.5rem 0;
  background-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)});
  box-shadow: 0 1px 1px 0 rgba(18, 18, 18, 0.2);
  border-radius: 1px;
  @apply --bbva-form-select-options-wrapper; }

.message {
  padding: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem) calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem) 0;
  display: flex;
  align-items: center;
  font-size: var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)});
  line-height: var(--lineHeightType3XSmall, ${unsafeCSS(foundations.lineHeight.type3XSmall)});
  color: var(--_message-color);
  @apply --bbva-form-select-message; }
  .message.info {
    color: var(--_info-message-color);
    @apply --bbva-form-select-info-message; }
  .message-icon {
    margin-right: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
    color: var(--_message-icon-color);
    @apply --bbva-form-select-message-icon; }
    .info .message-icon {
      color: var(--_info-message-icon-color);
      @apply --bbva-form-select-info-message-icon; }

@media (min-width: 600px) {
  .options-layer {
    position: absolute;
    top: 3rem;
    bottom: auto;
    display: block;
    padding: 0;
    transition: none;
    @apply --bbva-form-select-dt-options-layer; }
    .options-layer.opened {
      transition: none;
      @apply --bbva-form-select-dt-opened-options-layer; }
  .options-overlay {
    display: none;
    @apply --bbva-form-select-dt-options-overlay; }
  .options-gradient {
    max-width: none;
    @apply --bbva-form-select-dt-options-gradient; }
    .options-gradient::before, .options-gradient::after {
      display: none;
      @apply --bbva-form-select-dt-options-gradient-element; }
  .options-wrapper {
    padding: 0;
    max-height: none;
    min-height: 9rem;
    box-shadow: var(--boxShadowType1, ${unsafeCSS(foundations.boxShadow.type1)});
    @apply --bbva-form-select-dt-options-wrapper; } }
`;
