// @bbva-web-components/bbva-foundations-styles/bbva-foundations-styles-colors.js
export const colors = {
  primaryCore: '#004481', // dark background
  primaryCoreDarkened: '#072146',
  primaryCoreLightened: '#ffffff', // light background
  primaryCoreDark: '#043263',
  primaryCoreLight: '#1464A5', // medium background

  primaryMedium: '#1973B8',
  primaryMediumLight: '#49A5E6',

  primarySubdued: '#5BBEFF',
  primarySubduedLight: '#D4EDFC',

  primaryVariant: '#2DCCCD',
  primaryVariantDark: '#028484',
  primaryVariantMedium: '#02A5A5',
  primaryVariantLight: '#5AC4C4',
  primaryVariantLightened: '#EAF9FA',

  secondary600: '#121212',
  secondary500: '#666666',
  secondary400: '#BDBDBD',
  secondary300: '#D3D3D3',
  secondary200: '#E9E9E9',
  secondary100: '#F4F4F4',

  tertiaryType1: '#DA3851',
  tertiaryType1Dark: '#B92A45',
  tertiaryType1Medium: '#C0475E',
  tertiaryType1Light: '#E77D8E',
  tertiaryType1Lightened: '#F4C3CA',

  tertiaryType2: '#F7893B',
  tertiaryType2Dark: '#C65302',
  tertiaryType2Medium: '#D8732C',
  tertiaryType2Light: '#FAB27F',
  tertiaryType2Lightened: '#FDE7D8',

  tertiaryType3: '#F8CD51',
  tertiaryType3Dark: '#9C6C01',
  tertiaryType3Medium: '#C49735',
  tertiaryType3Light: '#FADE8E',
  tertiaryType3Lightened: '#FEF5DC',

  tertiaryType4: '#D8BE75',
  tertiaryType4Dark: '#8E7022',
  tertiaryType4Medium: '#B79E5E',
  tertiaryType4Light: '#E6D5A5',
  tertiaryType4Lightened: '#F3EBD5',

  tertiaryType5: '#F35E61',
  tertiaryType5Dark: '#CB353A',
  tertiaryType5Medium: '#D44B50',
  tertiaryType5Light: '#F59799',
  tertiaryType5Lightened: '#FCDFDF',

  tertiaryType6: '#48AE64',
  tertiaryType6Dark: '#277A3E',
  tertiaryType6Medium: '#388D4F',
  tertiaryType6Light: '#88CA9A',
  tertiaryType6Lightened: '#D9EFE0',

  tertiaryType7: '#F78BE8',
  tertiaryType7Dark: '#AD53A1',
  tertiaryType7Medium: '#C569B9',
  tertiaryType7Light: '#FAB3F0',
  tertiaryType7Lightened: '#FDDCF8',

  tertiaryType8: '#8F7AE5',
  tertiaryType8Dark: '#6754B8',
  tertiaryType8Medium: '#7C6AC7',
  tertiaryType8Light: '#B6A8EE',
  tertiaryType8Lightened: '#DDD7F7',
};
