// @bbva-web-components/bbva-clip-entities/bbva-clip-entities/3.3.0/bbva-clip-entities-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  display: block;
  box-sizing: border-box; }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

:host([size='small']),
:host(:not([size])) {
  width: 1.5rem;
  height: 1.5rem; }

:host([size='medium']) {
  width: 2rem;
  height: 2rem; }

:host([size='large']) {
  width: 2.5rem;
  height: 2.5rem; }

bbva-clip-box {
  --bbva-clip-box-bg-color: var(--bbva-clip-entities-clip-box-bg-color, var(--colorsPrimaryVariantDark, ${unsafeCSS(foundations.colors.primaryVariantDark)})); }
  bbva-clip-box[size='small'] {
    --bbva-clip-box-size: 1.5rem;
    --bbva-clip-box-s-icon-size: 0.75rem; }
  bbva-clip-box[size='medium'] {
    --bbva-clip-box-size: 2rem;
    --bbva-clip-box-m-icon-size: 1rem; }
  bbva-clip-box[size='large'] {
    --bbva-clip-box-size: 2.5rem;
    --bbva-clip-box-l-icon-size: 1.5rem; }

figure {
  margin: 0; }

.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

:host([disabled]) {
  opacity: var(--bbva-clip-entities-opacity, 0.4); }

:host([entity='bankia']) svg #forma1 {
  fill: var(--bbva-clip-entities-bankia-forma1-color, #BFD437); }

:host([entity='bankia']) svg #forma2 {
  fill: var(--bbva-clip-entities-bankia-forma2-color, #412613); }

:host([entity='bankinter']) svg #forma1 {
  fill: var(--bbva-clip-entities-bankinter-forma1-color, #F76A00); }

:host([entity='bankinter']) svg #forma2 {
  fill: var(--bbva-clip-entities-bankinter-forma2-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([entity='bbva']) svg #forma1 {
  fill: var(--bbva-clip-entities-bbva-forma1-color, #004284); }

:host([entity='bbva']) svg #forma2 {
  fill: var(--bbva-clip-entities-bbva-forma2-color, #0074BC); }

:host([entity='bbva']) svg #forma3 {
  fill: var(--bbva-clip-entities-bbva-forma3-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([entity='caixa']) svg #forma1 {
  fill: var(--bbva-clip-entities-caixa-forma1-color, #333333); }

:host([entity='kutxa']) svg #forma1 {
  fill: var(--bbva-clip-entities-kutxa-forma1-color, #000000); }

:host([entity='kutxa']) svg #forma2 {
  fill: var(--bbva-clip-entities-kutxa-forma2-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([entity='santander']) svg #forma1 {
  fill: var(--bbva-clip-entities-santander-forma1-color, #FD2415); }

:host([entity='santander']) svg #forma2 {
  fill: var(--bbva-clip-entities-santander-forma2-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([variant='white']) bbva-clip-box {
  --bbva-clip-box-bg-color: var(--bbva-clip-entities-white-clip-box-bg-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --bbva-clip-box-text-color: var(--bbva-clip-entities-white-clip-box-text-color, var(--colorsPrimaryVariantDark, ${unsafeCSS(foundations.colors.primaryVariantDark)})); }

:host([variant='white']) svg #forma1 {
  fill: var(--bbva-clip-entities-white-bg-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([variant='white']):host([entity='bankia']) svg #forma2 {
  fill: var(--bbva-clip-entities-white-bankia-forma2-color, #BFD437); }

:host([variant='white']):host([entity='bankinter']) svg #forma2 {
  fill: var(--bbva-clip-entities-white-bankinter-forma2-color, #FF6100); }

:host([variant='white']):host([entity='bbva']) svg #forma1 {
  fill: var(--bbva-clip-entities-white-bbva-forma1-color, var(--colorsSecondary100, ${unsafeCSS(foundations.colors.secondary100)})); }

:host([variant='white']):host([entity='bbva']) svg #forma2 {
  fill: var(--bbva-clip-entities-white-bbva-forma2-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([variant='white']):host([entity='bbva']) svg #forma3 {
  fill: var(--bbva-clip-entities-white-bbva-forma3-color, #004284); }

:host([variant='white']):host([entity='kutxa']) svg #forma2 {
  fill: var(--bbva-clip-entities-white-kutxa-forma2-color, #000000); }

:host([variant='white']):host([entity='santander']) svg #forma2 {
  fill: var(--bbva-clip-entities-white-santander-forma2-color, #FD2415); }
`;
