// @bbva-web-components/bbva-clip-entities/bbva-clip-entities/3.3.0/bbva-clip-entities.js
import { LitElement, html } from 'lit-element';
import { getComponentSharedStyles } from '@cells-components/cells-lit-helpers/cells-lit-helpers.js';
import * as entities from './bbva-clip-entities-logos.js';
import '@bbva-web-components/bbva-clip-box';
import styles from './bbva-clip-entities-styles.js';

export { bbvaClipEntitiesAmbient } from './bbva-clip-entities-ambient.js';

/**
The Entities clips are used to represent external companies this element is used in some components such as Button List,  List Movement and Panel Detail among others.
Clipboxes will be sized under the __three specific measurements__ (small 24px, medium 32px, large 40px) shown below and two color variation options in order to match over light or dark background compositions.

List entities:
* bankia
* bankinter
* bbva
* caixa
* kutxa
* santander

Example:

Default icon

```html
<bbva-clip-entities size="small"></bbva-clip-entities>
```

```html
<bbva-clip-entities size="small" entity="bbva">bbva</bbva-clip-entities>
```

``` html
<bbva-clip-entities entity="bbva" variant="white" size="large">bbva</bbva-clip-entities>
```

```html
<bbva-clip-entities entity="bbva" variant="white" size="large" disabled>bbva</bbva-clip-entities>
```

## Styling

The following custom properties are available for styling:

### Custom properties

| Selector          | CSS Property                                 | CSS Variable | Theme Variable               | Foundations/Fallback                                  |
| ----------------- | -------------------------------------------- | ------------ | ---------------------------- | ----------------------------------------------------- |
| [ambient^='dark'] | --bbva-clip-entities-white-bg-color          |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-white-bbva-forma1-color |              | --colorsSecondary100         | foundations.colors.secondary100         |
| [ambient^='dark'] | --bbva-clip-entities-white-bbva-forma2-color |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-bg-color                |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-bankia-forma1-color     |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-bankinter-forma1-color  |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-kutxa-forma1-color      |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-bbva-forma1-color       |              | --colorsSecondary100         | foundations.colors.secondary100         |
| [ambient^='dark'] | --bbva-clip-entities-bbva-forma2-color       |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-santander-forma1-color  |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-santander-forma1-color  |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-entities-caixa-forma1-color      |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='dark'] | --bbva-clip-box-bg-color                     |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
### Custom properties

| Selector           | CSS Property                                 | CSS Variable | Theme Variable               | Foundations/Fallback                                  |
| ------------------ | -------------------------------------------- | ------------ | ---------------------------- | ----------------------------------------------------- |
| [ambient^='light'] | --bbva-clip-entities-bankinter-forma2-color  |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-entities-bbva-forma3-color       |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-entities-kutxa-forma2-color      |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-entities-santander-forma2-color  |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-entities-white-bg-color          |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-entities-white-bbva-forma1-color |              | --colorsSecondary100         | foundations.colors.secondary100         |
| [ambient^='light'] | --bbva-clip-entities-white-bbva-forma2-color |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-box-text-color                   |              | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| [ambient^='light'] | --bbva-clip-box-bg-color                     |              | --colorsPrimaryVariantDark   | foundations.colors.primaryVariantDark   |
### Custom properties

| Selector                                                        | CSS Property               | CSS Variable                                      | Theme Variable               | Foundations/Fallback                                  |
| --------------------------------------------------------------- | -------------------------- | ------------------------------------------------- | ---------------------------- | ----------------------------------------------------- |
| :host([variant='white']):host([entity='santander']) svg #forma2 | fill                       | --bbva-clip-entities-white-santander-forma2-color |                              | #FD2415                                               |
| :host([variant='white']):host([entity='kutxa']) svg #forma2     | fill                       | --bbva-clip-entities-white-kutxa-forma2-color     |                              | #000000                                               |
| :host([variant='white']):host([entity='bbva']) svg #forma3      | fill                       | --bbva-clip-entities-white-bbva-forma3-color      |                              | #004284                                               |
| :host([variant='white']):host([entity='bbva']) svg #forma2      | fill                       | --bbva-clip-entities-white-bbva-forma2-color      | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([variant='white']):host([entity='bbva']) svg #forma1      | fill                       | --bbva-clip-entities-white-bbva-forma1-color      | --colorsSecondary100         | foundations.colors.secondary100         |
| :host([variant='white']):host([entity='bankinter']) svg #forma2 | fill                       | --bbva-clip-entities-white-bankinter-forma2-color |                              | #FF6100                                               |
| :host([variant='white']):host([entity='bankia']) svg #forma2    | fill                       | --bbva-clip-entities-white-bankia-forma2-color    |                              | #BFD437                                               |
| :host([variant='white']) svg #forma1                            | fill                       | --bbva-clip-entities-white-bg-color               | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([variant='white']) bbva-clip-box                          | --bbva-clip-box-bg-color   | --bbva-clip-entities-white-clip-box-bg-color      | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([variant='white']) bbva-clip-box                          | --bbva-clip-box-text-color | --bbva-clip-entities-white-clip-box-text-color    | --colorsPrimaryVariantDark   | foundations.colors.primaryVariantDark   |
| :host([entity='santander']) svg #forma2                         | fill                       | --bbva-clip-entities-santander-forma2-color       | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([entity='santander']) svg #forma1                         | fill                       | --bbva-clip-entities-santander-forma1-color       |                              | #FD2415                                               |
| :host([entity='kutxa']) svg #forma2                             | fill                       | --bbva-clip-entities-kutxa-forma2-color           | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([entity='kutxa']) svg #forma1                             | fill                       | --bbva-clip-entities-kutxa-forma1-color           |                              | #000000                                               |
| :host([entity='caixa']) svg #forma1                             | fill                       | --bbva-clip-entities-caixa-forma1-color           |                              | #333333                                               |
| :host([entity='bbva']) svg #forma3                              | fill                       | --bbva-clip-entities-bbva-forma3-color            | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([entity='bbva']) svg #forma2                              | fill                       | --bbva-clip-entities-bbva-forma2-color            |                              | #0074BC                                               |
| :host([entity='bbva']) svg #forma1                              | fill                       | --bbva-clip-entities-bbva-forma1-color            |                              | #004284                                               |
| :host([entity='bankinter']) svg #forma2                         | fill                       | --bbva-clip-entities-bankinter-forma2-color       | --colorsPrimaryCoreLightened | foundations.colors.primaryCoreLightened |
| :host([entity='bankinter']) svg #forma1                         | fill                       | --bbva-clip-entities-bankinter-forma1-color       |                              | #F76A00                                               |
| :host([entity='bankia']) svg #forma2                            | fill                       | --bbva-clip-entities-bankia-forma2-color          |                              | #412613                                               |
| :host([entity='bankia']) svg #forma1                            | fill                       | --bbva-clip-entities-bankia-forma1-color          |                              | #BFD437                                               |
| :host([disabled])                                               | opacity                    | --bbva-clip-entities-opacity                      |                              | 0.4                                                   |
| bbva-clip-box                                                   | --bbva-clip-box-bg-color   | --bbva-clip-entities-clip-box-bg-color            | --colorsPrimaryVariantDark   | foundations.colors.primaryVariantDark   |
> Styling documentation generated by Cells CLI

@customElement bbva-clip-entities
@polymer
@LitElement
@demo demo/index.html
*/
export class BbvaClipEntities extends LitElement {
  static get is() {
    return 'bbva-clip-entities';
  }

  static get properties() {
    return {
      /**
       * Disabled state
       */
      disabled: {
        type: Boolean,
        reflect: true,
      },
      /**
       * The name of the entity to use.
       */
      entity: {
        type: String,
      },
      /**
       * Entity  size
       * Values: "small", "medium" or "large"
       */
      size: {
        type: String,
        reflect: true,
      },
      /**
       * Default entity icon
       * @default 'coronita:bank'
       */
      defaultEntityIcon: {
        type: String,
        attribute: 'default-entity-icon',
      },
    };
  }

  constructor() {
    super();
    this.entity = '';
    this.disabled = false;
    this.defaultEntityIcon = 'coronita:bank';
  }

  static get styles() {
    return [styles, getComponentSharedStyles('bbva-clip-entities-shared-styles')];
  }

  render() {
    return html`
      ${this._entityLogo
        ? html`
            <span class="sr-only"><slot></slot></span>
            <figure>${this._entityLogo}</figure>
          `
        : html`
            <bbva-clip-box size="${this.size}" .icon="${this.defaultEntityIcon}"></bbva-clip-box>
          `}
    `;
  }

  get _entityLogo() {
    const entityName = this.entity.toLowerCase();
    return entities[entityName];
  }
}

customElements.define(BbvaClipEntities.is, BbvaClipEntities);
