// @cells-components/cells-demo-resizable-viewport/cells-demo-resizable-viewport-styles.js
import { css, } from 'lit-element';

export default css`
:host {
  --cells-demo-resizable-viewport-background-color: #f0f0f0;
  --cells-demo-resizable-viewport-dragger-color: #A9ABAD;
  display: block;
  font-family: sans-serif; }

*, *::before, *::after {
  font: inherit;
  box-sizing: inherit; }

.bp {
  height: 26px;
  box-sizing: border-box;
  position: relative;
  display: flex;
  justify-content: flex-end;
  align-items: center;
  @apply --cells-demo-resizable-viewport-bar-wrapper; }
  .bp__bar {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
    background-color: #D0D1D4;
    color: #fff;
    border: 4px solid var(--cells-demo-resizable-viewport-background-color);
    border-top: 0;
    box-sizing: content-box;
    font-size: 0.625rem;
    letter-spacing: 1px;
    display: flex;
    align-items: center;
    justify-content: center;
    cursor: pointer;
    overflow: auto;
    -webkit-user-select: none;
       -moz-user-select: none;
        -ms-user-select: none;
            user-select: none;
    @apply --cells-demo-resizable-viewport-bar; }
    .bp__bar .dragger {
      content: "";
      width: 10px;
      height: 10px;
      position: absolute;
      right: 2px;
      bottom: 2px;
      @apply --cells-demo-resizable-viewport-dragger; }
    .bp__bar .dragger-path {
      fill: var(--cells-demo-resizable-viewport-dragger-inactive-color, #fff); }
    .bp__bar.cells-selected {
      background-color: #343B4A;
      color: #fff;
      @apply --cells-demo-resizable-viewport-bar-selected; }
      .bp__bar.cells-selected ~ .bp__bar {
        background-color: transparent;
        @apply --cells-demo-resizable-viewport-bar-selected-siblings; }
        .bp__bar.cells-selected ~ .bp__bar .dragger-path {
          fill: var(--cells-demo-resizable-viewport-dragger-color); }
        .bp__bar.cells-selected ~ .bp__bar:hover {
          background-color: #252a35;
          @apply --cells-demo-resizable-viewport-bar-selected-siblings-hover; }
      .bp__bar.cells-selected .dragger-path {
        fill: var(--cells-demo-resizable-viewport-dragger-color); }
    .bp__bar:hover {
      background-color: #252a35;
      @apply --cells-demo-resizable-viewport-bar-hover; }
      .bp__bar:hover ~ .bp__bar {
        background-color: transparent;
        @apply --cells-demo-resizable-viewport-bar-hover-siblings; }
        .bp__bar:hover ~ .bp__bar .dragger-path {
          fill: var(--cells-demo-resizable-viewport-dragger-color); }
      .bp__bar:hover .bp__label {
        visibility: visible;
        position: relative;
        background-color: #252a35;
        @apply --cells-demo-resizable-viewport-bar-hover-label; }
      .bp__bar:hover ~ .bp__bar .bp__label {
        visibility: hidden;
        @apply --cells-demo-resizable-viewport-bar-hover-label-siblings; }
      .bp__bar:hover .dragger-path {
        fill: var(--cells-demo-resizable-viewport-dragger-color); }
  .bp__label {
    text-transform: uppercase;
    visibility: hidden;
    background-color: #252a35;
    @apply --cells-demo-resizable-viewport-bar-label; }
    .cells-selected .bp__label {
      visibility: visible;
      background-color: #343B4A;
      @apply --cells-demo-resizable-viewport-bar-selected-label; }

.btn-reset {
  font-family: inherit;
  font-size: 0.75rem;
  background-color: rgba(0, 0, 0, 0.6);
  border: none;
  color: #fff;
  border-radius: 3px;
  padding: 4px 8px;
  cursor: pointer;
  outline: none;
  position: relative;
  z-index: 9;
  -webkit-user-select: none;
     -moz-user-select: none;
      -ms-user-select: none;
          user-select: none;
  @apply --cells-demo-resizable-viewport-button-reset; }
  .btn-reset:hover, .btn-reset:active, .btn-reset:focus {
    background-color: rgba(0, 0, 0, 0.8);
    @apply --cells-demo-resizable-viewport-button-reset-active; }
`;
