// @bbva-web-components/bbva-amount/bbva-amount/bbva-amount-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles/bbva-foundations-styles.js';

export default css`:host {
  display: inline-block;
  box-sizing: border-box;
  vertical-align: bottom;
  max-width: 100%;
  font-weight: var(--bbva-amount-font-weight, var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(foundations.fontFacePrimary.book.fontWeight)})); }

*,
*:before,
*:after {
  box-sizing: inherit; }

.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

.amount-wrapper {
  display: block;
  word-spacing: -0.25em;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  unicode-bidi: bidi-override; }

.amount-block {
  margin: 0 0.125rem; }

@supports (-moz-appearance: meterbar) {
  .amount-wrapper {
    word-spacing: calc(-0.25em + 1.35px); } }

.minus-left {
  float: var(--bbva-amount-minus-left-float, none); }

.minus-right {
  float: var(--bbva-amount-minus-right-float, none); }

.currency {
  float: var(--bbva-amount-currency-float, none); }
  .currency-left, .currency-right {
    vertical-align: top; }

.currency-code {
  word-spacing: normal; }

.abbreviation {
  word-spacing: normal; }

:host([hidden]),
[hidden] {
  display: none !important; }

:host([variant='l']) {
  font-size: var(--bbva-amount-l-font-size, var(--typographyTypeLarge, ${unsafeCSS(foundations.typography.typeLarge)}));
  line-height: var(--bbva-amount-l-line-height, var(--lineHeightTypeLarge, ${unsafeCSS(foundations.lineHeight.typeLarge)})); }
  :host([variant='l']) .currency {
    font-size: var(--bbva-amount-l-currency-font-size, var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)}));
    line-height: var(--bbva-amount-l-currency-line-height, var(--lineHeightType3XSmall, ${unsafeCSS(foundations.lineHeight.type3XSmall)})); }

:host([variant='xl']) {
  font-size: var(--bbva-amount-xl-font-size, var(--typographyTypeXLarge, ${unsafeCSS(foundations.typography.typeXLarge)}));
  line-height: var(--bbva-amount-xl-line-height, var(--lineHeightTypeXLarge, ${unsafeCSS(foundations.lineHeight.typeXLarge)})); }
  :host([variant='xl']) .currency {
    font-size: var(--bbva-amount-xl-currency-font-size, var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)}));
    line-height: var(--bbva-amount-xl-currency-line-height, var(--lineHeightType3XSmall, ${unsafeCSS(foundations.lineHeight.type3XSmall)})); }

:host([variant='3xl']) {
  font-size: var(--bbva-amount-3xl-font-size, var(--typographyType3XLarge, ${unsafeCSS(foundations.typography.type3XLarge)}));
  line-height: var(--bbva-amount-3xl-line-height, var(--lineHeightType3XLarge, ${unsafeCSS(foundations.lineHeight.type3XLarge)})); }
  :host([variant='3xl']) .currency {
    font-size: var(--bbva-amount-3xl-currency-font-size, var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)}));
    line-height: var(--bbva-amount-3xl-currency-line-height, var(--lineHeightType3XSmall, ${unsafeCSS(foundations.lineHeight.type3XSmall)})); }

:host([variant='4xl']) {
  font-size: var(--bbva-amount-4xl-font-size, 2rem);
  line-height: var(--bbva-amount-4xl-line-height, 2rem); }
  :host([variant='4xl']) .currency {
    font-size: var(--bbva-amount-4xl-currency-font-size, var(--typographyTypeXLarge, ${unsafeCSS(foundations.typography.typeXLarge)}));
    line-height: var(--bbva-amount-4xl-currency-line-height, var(--lineHeightTypeXLarge, ${unsafeCSS(foundations.lineHeight.typeXLarge)})); }

:host([variant='5xl']) {
  font-size: var(--bbva-amount-5xl-font-size, 3rem);
  line-height: var(--bbva-amount-5xl-line-height, 3rem); }
  :host([variant='5xl']) .currency {
    font-size: var(--bbva-amount-5xl-currency-font-size, 1.75rem);
    line-height: var(--bbva-amount-5xl-currency-line-height, 2rem); }
`;
