// @bbva-web-components/bbva-form-select/bbva-form-option-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  display: flex;
  align-items: center;
  box-sizing: border-box;
  padding: 0 calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
  font-size: var(--typographyTypeSmall, ${unsafeCSS(foundations.typography.typeSmall)});
  line-height: var(--lineHeightTypeSmall, ${unsafeCSS(foundations.lineHeight.typeSmall)});
  font-weight: var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(foundations.fontFacePrimary.book.fontWeight)});
  font-style: var(--fontFacePrimaryBookFontStyle, ${unsafeCSS(foundations.fontFacePrimary.book.fontStyle)});
  color: var(--bbva-form-option-color, var(--colorsSecondary600, ${unsafeCSS(foundations.colors.secondary600)}));
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  cursor: pointer;
  outline: none;
  @apply --bbva-form-option; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *::before, *::after {
  box-sizing: inherit; }

.clip {
  flex: none;
  @apply --bbva-form-option-clip; }
  .clip ::slotted(*) {
    margin-right: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
    @apply --bbva-form-option-clip-slotted; }

.text {
  flex: auto;
  margin: 0.75rem 0;
  min-width: 0;
  white-space: normal;
  @apply --bbva-form-option-text; }

.icon {
  flex: none;
  margin-left: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
  @apply --bbva-form-option-icon; }

:host([tabindex='0']) {
  position: relative;
  background-color: var(--colorsPrimaryCore, ${unsafeCSS(foundations.colors.primaryCore)});
  color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)});
  z-index: 1;
  @apply --bbva-form-option-focused; }

:host([aria-disabled=true]) {
  opacity: .4;
  cursor: default;
  @apply --bbva-form-option-disabled; }
`;
