// @cells-components/cells-demo-event-toaster/cells-demo-event-toaster-styles.js
import { css, } from 'lit-element';

export default css`
:host {
  @apply --cells-demo-event-toaster; }

*, *::before, *::after {
  font: inherit;
  box-sizing: inherit; }

:host(.full-width) paper-toast {
  width: 100%;
  min-width: 0;
  border-radius: 0;
  margin: 0; }

.highlight {
  font-family: Consolas, Menlo, Monaco, monospace;
  font-size: 14px;
  font-size: inherit;
  color: var(--cells-demo-event-toaster-highlight-color, yellow);
  @apply --cells-demo-event-toaster-highlight; }

paper-toast {
  font-family: sans-serif;
  letter-spacing: normal;
  background: transparent;
  box-shadow: none;
  padding: 0 !important;
  display: flex;
  flex-direction: column-reverse;
  /* display toasts in reverse order without reversing the array */ }

:host([toast-vertical-align='top']) paper-toast {
  margin-top: var(--cells-demo-event-toaster-vertical-top, 160px); }

p {
  margin: 0; }

.inner-toast {
  display: none;
  background: #323232;
  color: #f1f1f1;
  border-radius: 2px;
  box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.26);
  font-size: 14px;
  padding: 15px;
  margin-top: 10px; }

[visible] {
  display: block; }
`;
