// @cells-components/cells-demo-helper/cells-demo-helper-styles.js
import { css } from 'lit-element';

export default css`:host {
  --cells-demo-helper-text-color: #343B4A;
  display: block;
  font-family: var(--cells-fontDefault, Lato, sans-serif);
  color: var(--cells-demo-helper-text-color);
  box-sizing: border-box;
  min-height: 100vh;
  background: radial-gradient(rgba(171, 171, 171, 0.5) 5.5%, transparent 6.5%), #f0f0f0;
  background-size: 24px 24px;
  background-position: 28px 20px;
  @apply --cells-demo-helper; }
  @media (min-width: 1024px) {
    :host {
      overflow-x: hidden; } }

:host([hidden]), [hidden] {
  display: none; }

*, *:before, *:after {
  box-sizing: inherit;
  font: inherit; }

:host([hide-ui]) .header,
:host([hide-ui]) .resizer {
  display: none; }

.ui-checkbox {
  position: absolute;
  z-index: 9;
  top: 10px;
  right: 10px;
  @apply --cells-demo-helper-ui-checkbox; }
  .ui-checkbox__label {
    display: flex;
    align-items: center;
    font-size: 0.875rem; }
    @media (min-width: 768px) {
      :host(:not([hide-ui])) .ui-checkbox__label#hideUI {
        display: none; } }
  .ui-checkbox__input {
    margin: -1px 0.375rem 0 0;
    font-size: 1.125rem; }

.header {
  padding: 20px 24px 36px;
  background-color: #fff;
  @apply --cells-demo-helper-header; }
  @media (min-width: 1024px) {
    .header {
      padding: 36px 24px; } }

.tabs {
  max-width: 200px;
  font-size: 0.875rem;
  font-weight: bold;
  letter-spacing: 1px;
  margin-left: -10px;
  @apply --cells-demo-helper-tabs;
  --cells-demo-radio-tabs-color-selected: var(--cells-demo-helper-text-color);
  --cells-demo-radio-tabs-border-bottom-color: transparent;
  --cells-demo-radio-tabs-item-height: rem(40px);
  --cells-demo-radio-tabs-content: {
    border: 2px solid transparent;
    border-radius: 40px;
    display: block;
    line-height: 36px;
    padding-left: 15px;
    padding-right: 15px;
  };
  --cells-demo-radio-tabs-indicator: {
    display:none;
  };
  --cells-demo-radio-tabs-selected: {
    border: 2px solid currentColor;
  };
  --cells-demo-radio-tabs-option: {
    margin: 0 9px;
  }; }
  @media (max-width: 630px) {
    .tabs {
      order: 1;
      margin-left: 0;
      margin-top: 20px; } }
  @media (min-width: 1024px) {
    .tabs {
      position: absolute;
      top: 15px;
      left: 0;
      right: 0;
      margin: auto; } }

.top-row {
  display: flex;
  justify-content: space-between;
  max-width: 1024px;
  margin: auto;
  position: relative;
  align-items: center;
  flex-wrap: wrap;
  @apply --cells-demo-helper-top; }
  @media (max-width: 630px) {
    .top-row {
      flex-direction: column; } }
  @media (min-width: 1024px) {
    .top-row {
      flex-wrap: nowrap; } }

.heading {
  flex: 1 0 100%;
  margin-bottom: 20px;
  color: inherit;
  @apply --cells-demo-helper-case-heading; }
  @media (max-width: 630px) {
    .heading {
      width: 100%; } }
  @media (min-width: 1024px) {
    .heading {
      margin-bottom: -10px;
      max-width: 400px;
      flex: 0 0 auto; } }
  .heading-title {
    margin: 0;
    font-weight: 300;
    font-size: 1.25rem;
    text-transform: uppercase;
    @apply --cells-demo-helper-case-heading-title; }
  .heading-description {
    margin: 6px 0 0;
    font-weight: 300;
    font-size: 0.8125rem;
    @apply --cells-demo-helper-case-heading-description; }

.selectors {
  display: flex;
  margin-left: auto;
  position: relative;
  z-index: 10;
  @apply --cells-demo-helper-selectors; }
  @media (max-width: 630px) {
    .selectors {
      margin-left: 0;
      width: 100%;
      flex-direction: column; } }

.dropdown {
  margin-left: 1.5rem;
  width: 7.5rem;
  z-index: 3;
  @apply --cells-demo-helper-dropdown; }
  @media (max-width: 630px) {
    .dropdown {
      width: 100%;
      margin-left: 0;
      z-index: initial; } }

@media (min-width: 631px) {
  .lang-selector {
    width: 80px; } }

.main {
  position: relative;
  padding: 80px 0;
  @apply --cells-demo-helper-main; }
  @media (max-width: 767px) {
    .main {
      padding: 0; } }

.resizer {
  padding-top: 4px;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  margin: auto;
  overflow: hidden;
  @apply --cells-demo-helper-resize-ui;
  --cells-demo-resizable-viewport-button-reset: {
    margin-right: 10px;
    margin-top: -2px;
  }; }
  @media (max-width: 767px) {
    .resizer {
      display: none;
      @apply --cells-demo-helper-small-resize-ui; } }

.iframe-container {
  will-change: width, height;
  contain: size;
  width: 360px;
  margin: 0 auto;
  background-color: #fff;
  box-sizing: content-box;
  position: relative;
  transition: width 0.2s ease-out, height 0.2s ease-out;
  @apply --cells-demo-helper-iframe-container; }
  @media (max-width: 767px) {
    .iframe-container {
      margin: 0;
      width: 100% !important;
      height: auto;
      overflow: hidden;
      min-height: 100vh;
      @apply --cells-demo-helper-small-iframe-container; } }
  @media (min-width: 768px) {
    .iframe-container {
      min-height: 360px;
      @apply --cells-demo-helper-medium-iframe-container; } }
  .iframe-container iframe {
    width: 100%;
    height: 100%;
    border: none;
    position: relative;
    background-color: #fff;
    z-index: 1;
    @apply --cells-demo-helper-iframe; }
  .iframe-container:before {
    content: '';
    display: block;
    position: absolute;
    top: -16px;
    left: -16px;
    right: -16px;
    bottom: -16px;
    z-index: 0;
    background-color: rgba(52, 59, 74, 0.2);
    border-radius: 20px;
    @apply --cells-demo-helper-iframe-mobile-frame; }

.overlay {
  visibility: hidden;
  opacity: 0;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  padding: 0 20px;
  height: 100%;
  z-index: 2;
  background-color: #f0f0f0;
  border-top: 1px solid #dcdcdc;
  background: linear-gradient(to bottom, #f0f0f0 1%, rgba(240, 240, 240, 0.7) 90%, rgba(240, 240, 240, 0.5) 100%);
  transition: visibility 0s linear 120ms, opacity 120ms linear;
  will-change: visibility, opacity;
  @apply --cells-demo-helper-overlay; }
  .overlay.visible {
    visibility: visible;
    opacity: 1;
    transition-delay: 0s;
    @apply --cells-demo-helper-overlay-visible; }

.code-wrapper {
  opacity: 0;
  visibility: hidden;
  height: 100%;
  overflow: hidden;
  @apply --cells-demo-helper-code-wrapper; }

.visible .code-wrapper {
  visibility: visible;
  opacity: 1;
  transition-delay: 0s;
  @apply --cells-demo-helper-code-wrapper-visible; }

.overlay-inner {
  position: relative;
  height: 100%;
  max-width: 1024px;
  margin: 0 auto;
  @apply --cells-demo-helper-code-overlay-inner; }

.disclaimer {
  margin: 1rem 0; }
  .disclaimer span {
    letter-spacing: 2px;
    line-height: 26px; }
    .disclaimer span:first-child {
      display: inline-block;
      text-transform: uppercase;
      padding-left: 0.2em;
      font-weight: 300;
      font-size: 1.32rem; }
    .disclaimer span + span {
      display: block;
      color: rgba(52, 59, 73, 0.84);
      text-indent: 0.3rem;
      font-size: 0.88rem; }

.copy-button {
  display: block;
  position: absolute;
  z-index: 1;
  top: 12px;
  right: 0.2rem;
  color: #fff;
  background-color: #3D474E;
  border: 0;
  font-size: 10px;
  font-weight: bold;
  text-transform: uppercase;
  letter-spacing: 2px;
  border-radius: 20px;
  line-height: 40px;
  padding: 0 15px;
  @apply --cells-demo-helper-code-copy-button; }

.code {
  font-size: 14px;
  @apply --cells-demo-helper-code; }

.code > pre {
  margin: 0;
  @apply --cells-demo-helper-code-pre; }
`;
