// @bbva-web-components/bbva-header-main/bbva-header-main/3.2.1/bbva-header-main-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  display: block;
  box-sizing: border-box;
  background-color: var(--bbva-header-main-bg-color, var(--backgroundColorsDark400, ${unsafeCSS(foundations.backgroundColors.dark400)}));
  color: var(--bbva-header-main-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

.header-main {
  display: flex;
  justify-content: space-between;
  align-items: center;
  position: relative;
  padding-top: var(--bbva-header-main-padding-top, 0);
  padding-left: calc(((0.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
  padding-right: calc(((0.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem);
  margin-left: auto;
  margin-right: auto;
  min-height: var(--bbva-header-main-min-height, calc(((6 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
  max-width: var(--bbva-header-main-max-width, 100%); }
  .header-main .text {
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    display: block;
    position: absolute;
    bottom: 0;
    top: var(--bbva-header-main-text-top, var(--bbva-header-main-padding-top, 0));
    left: 0;
    width: 100%;
    height: var(--bbva-header-main-min-height, calc(((6 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
    margin: 0;
    padding: 0 var(--bbva-header-main-text-padding-horizontal, calc(((6 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
    text-align: center;
    font-size: var(--bbva-header-main-text-font-size, var(--typographyTypeMedium, ${unsafeCSS(foundations.typography.typeMedium)}));
    line-height: var(--bbva-header-main-text-line-height, calc(((6 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
    font-weight: 400; }
    .header-main .text.full-padding {
      --bbva-header-main-text-padding-horizontal: calc(((12 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem); }
    .header-main .text.image {
      display: flex;
      justify-content: center; }
  .header-main .header-icons {
    display: flex; }
  .header-main .icon {
    display: block;
    position: relative;
    border: none;
    background: none;
    width: var(--bbva-header-main-min-height, calc(((5.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
    min-height: var(--bbva-header-main-min-height, calc(((6 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem));
    padding: 0;
    color: inherit;
    margin: 0; }

:host(.modal) {
  --bbva-header-main-bg-color: var(--backgroundColorsDark, ${unsafeCSS(foundations.backgroundColors.dark)}); }

:host(.modal-three-level) {
  --bbva-header-main-bg-color: var(--backgroundColorsLight, ${unsafeCSS(foundations.backgroundColors.light)});
  --bbva-header-main-color: var(--colorsSecondary600, ${unsafeCSS(foundations.colors.secondary600)}); }

:host(.process) {
  --bbva-header-main-bg-color: var(--colorsTertiaryType6Dark, ${unsafeCSS(foundations.colors.tertiaryType6Dark)}); }

:host(.sign) {
  --bbva-header-main-bg-color: var(--colorsTertiaryType4Dark, ${unsafeCSS(foundations.colors.tertiaryType4Dark)}); }

:host(.transparent) {
  --bbva-header-main-bg-color: transparent; }
`;
