// @cells-components/cells-demo-dropdown-menu/cells-demo-dropdown-menu/cells-demo-dropdown-menu-styles.js
import { css } from 'lit-element';

export default css`:host {
  display: inline-block;
  @apply --cells-demo-dropdown-menu; }

*, *::before, *::after {
  font: inherit;
  box-sizing: inherit; }

:host([hidden]), [hidden] {
  display: none !important; }

.label {
  font-size: 0.75rem;
  margin-bottom: 0.625rem;
  color: var(--cells-demo-grey-2);
  letter-spacing: 1px;
  @apply --cells-demo-dropdown-menu-label; }

.dropdown {
  width: 100%;
  letter-spacing: 1px;
  @apply --cells-demo-dropdown-menu-dropdown;
  --cells-dropdown-layer-hovered-underline-color: var(--cells-demo-grey-4);
  --cells-dropdown-layer-opened-underline-color: var(--cells-demo-grey-4);
  --cells-dropdown-layer-hovered-icon-color: var(--cells-demo-grey-4);
  --cells-dropdown-layer-opened-icon-color: var(--cells-demo-grey-4);
  --cells-dropdown-menu-value-color: var(--cells-demo-grey-1);
  --cells-dropdown-layer-control-text: unquote("{
    font-size: 0.875rem;
    color: var(--cells-demo-grey-3);
  }");
  --cells-dropdown-menu-control-icon-check: unquote("{
    width: 1rem;
    height: 1rem;
    margin-top: -0.1875rem;
  }");
  --cells-dropdown-menu-control-icon-check-checked: unquote("{
    color: var(--cells-demo-menu-control-green, #1DFD92);
  }");
  --cells-dropdown-layer-line-icon: unquote("{
    width: 1rem;
    height: 1rem;
  }");
  --cells-dropdown-layer-line-icons: unquote("{
    bottom: -3px;
  }");
  --cells-dropdown-layer-line-icon-open: unquote("{
    transform-origin: center;
  }");
  --cells-dropdown-layer-opened-line-icon-open: unquote("{
    transform-origin: center;
    transform: rotate(180deg) translateY(0.75rem);
  }");
  --cells-dropdown-layer-hovered-line-underline-after: unquote("{
    transform: translateX(-1rem);
  }"); }
`;
