// @cells-components/cells-demo-radio-tabs/cells-demo-radio-tabs-styles.js
import { css } from 'lit-element';

export default css`:host {
  --cells-demo-radio-tabs-border-radius: 0;
  --cells-demo-radio-tabs-item-width: 100%;
  display: flex;
  position: relative;
  font-size: 0.8125rem;
  font-weight: 500;
  color: var(--cells-demo-radio-tabs-color, var(--bbva-300, #D3D3D3));
  height: var(--cells-demo-radio-tabs-item-height, 3.4375rem);
  background-color: var(--cells-demo-radio-tabs-bg-color, var(--bbva-white, #fff));
  border-radius: var(--cells-demo-radio-tabs-border-radius);
  box-sizing: border-box;
  @apply --cells-demo-radio-tabs; }

label {
  display: flex;
  justify-content: center;
  align-items: center;
  position: relative;
  flex: var(--cells-demo-radio-tabs-option-flex, 1);
  height: 100%;
  cursor: pointer;
  border-bottom: 1px solid var(--cells-demo-radio-tabs-border-bottom-color, var(--bbva-200, #E9E9E9));
  outline: none;
  @apply --cells-demo-radio-tabs-option; }

label:focus .tab-content {
  color: var(--cells-demo-radio-tabs-color-selected, var(--bbva-core-blue, #004481));
  @apply --cells-demo-radio-tabs-option-focus; }

label:hover .tab-content {
  @apply --cells-demo-radio-tabs-tab-hover; }

input {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

input[checked] + .tab-content {
  color: var(--cells-demo-radio-tabs-color-selected, var(--bbva-core-blue, #004481));
  @apply --cells-demo-radio-tabs-selected; }

.tab-content {
  display: flex;
  align-items: center;
  pointer-events: none;
  @apply --cells-demo-radio-tabs-content; }

.icon {
  margin: var(--cells-demo-radio-tabs-icon-margin, 0 0.625rem 0 0);
  @apply --cells-demo-radio-tabs-icon; }

#indicator {
  position: absolute;
  left: 0;
  bottom: 0;
  width: var(--cells-demo-radio-tabs-item-width);
  height: 0.125rem;
  transition: transform 0.2s;
  border-radius: var(--cells-demo-radio-tabs-border-radius);
  background-color: var(--cells-demo-radio-tabs-indicator-bg-color, var(--bbva-core-blue, #004481));
  @apply --cells-demo-radio-tabs-indicator; }

:host(.indicator-hidden) #indicator {
  opacity: 0; }

:host([hidden]), [hidden] {
  display: none; }

*, *:before, *:after {
  box-sizing: inherit;
  font: inherit; }
`;
