// @bbva-web-components/bbva-list-bullet/bbva-list-bullet/bbva-list-bullet-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles/bbva-foundations-styles.js';

export default css`:host {
  display: block;
  box-sizing: border-box;
  position: relative;
  padding: 0 0 calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1em) calc(((3 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1em);
  color: var(--bbva-list-bullet-color, var(--bbva-600, #121212));
  font-size: var(--typographyTypeSmall, ${unsafeCSS(foundations.typography.typeSmall)});
  line-height: var(--lineHeightTypeSmall, ${unsafeCSS(foundations.lineHeight.typeSmall)}); }

.icon {
  position: absolute;
  top: 0;
  left: 0;
  color: var(--bbva-list-bullet-icon-color, var(--bbva-medium-blue, #1973b8)); }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

:host(.transparent) {
  background: none;
  --bbva-list-bullet-icon-color: var(--bbva-aqua, #2dcccd);
  --bbva-list-bullet-color: var(--bbva-white, #ffffff); }
`;
