// @bbva-web-components/bbva-clip-product/bbva-clip-product/4.0.1/bbva-clip-product-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  --_bg-color: var(--bbva-clip-product-bg-color, var(--colorsSecondary100, ${unsafeCSS(foundations.colors.secondary100)}));
  --_secondary-bg-color: var(--bbva-clip-product-secondary-bg-color,var(--colorsSecondary200, ${unsafeCSS(foundations.colors.secondary200)}));
  display: inline-block;
  box-sizing: border-box; }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

.bg {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 4rem;
  height: 4rem;
  border-radius: 6.25rem;
  background-color: var(--_bg-color); }
  .bg.large {
    width: 3rem;
    height: 3rem; }

.item {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 2.125rem;
  height: 2.125rem; }

.avatar {
  display: flex;
  justify-content: center;
  align-items: center; }

svg {
  width: 100%;
  height: 100%; }

:host([variant='maximum']) .bg {
  width: 8.5rem;
  height: 4rem;
  background-image: linear-gradient(110deg, var(--_secondary-bg-color) 50%, transparent 0); }

:host([variant='maximum']) .item {
  flex-basis: 50%;
  width: 4rem;
  height: 2rem; }

:host([variant='maximum']) .avatar {
  flex-basis: 50%; }

:host([item='card']) .large .item {
  height: 1.5rem; }

:host([item='phone']) .large .item {
  height: 2.25rem; }
`;
