// @bbva-web-components/bbva-list-bullet/bbva-list-bullet.js
import { LitElement, html } from 'lit-element';
import { getComponentSharedStyles } from '@cells-components/cells-lit-helpers';
import '@cells-components/cells-icon/cells-icon.js';
import styles from './bbva-list-bullet-styles.js';
/**
Text with a bullet icon such as an element of list with a dash

Example:
```html
<bbva-list-bullet>hello  item</bbva-list-bullet>
```

To reset the html list styles use the bbva-list-bullet-ul-styles.js included in the component
```html

<template>
  <ul></ul>
  <script type="module">
      import styles from '../bbva-list-bullet-ul-styles.js';
      import {html, render} from '../../../../node_modules/lit-html/lit-html.js';
      const items = ['blue', 'red'];
      const bulletTemplate = (items) => html`
        <style>styles,</style>
        ${items.map((item, i) => html`<li><bbva-list-bullet>${item}</bbva-list-bullet></li>`)}
      `;
      render(bulletTemplate(items), document.querySelector('ul'));
  </script>
</template>
```

## Icons

Since this component uses icons, it will need an [iconset](https://platform.bbva.com/en-us/developers/engines/cells/documentation/cells-architecture/composing-with-components/cells-icons) in your project as an application level dependency. In fact, this component uses an iconset in its demo.

## Styling

The following custom properties are available for styling:

### Custom properties

| Selector            | CSS Property                  | CSS Variable                  | Theme Variable        | Foundations/Fallback                           |
| ------------------- | ----------------------------- | ----------------------------- | --------------------- | ---------------------------------------------- |
| :host(.transparent) | --bbva-list-bullet-icon-color | --bbva-aqua                   |                       | #2dcccd                                        |
| :host(.transparent) | --bbva-list-bullet-color      | --bbva-white                  |                       | #ffffff                                        |
| .icon               | color                         | --bbva-list-bullet-icon-color | --bbva-medium-blue    | #1973b8                                        |
| :host               | padding                       | --gridSpacerVariant           | --gridSpacer          | foundations.grid.spacer          |
| :host               | color                         | --bbva-list-bullet-color      | --bbva-600            | #121212                                        |
| :host               | font-size                     |                               | --typographyTypeSmall | foundations.typography.typeSmall |
| :host               | line-height                   |                               | --lineHeightTypeSmall | foundations.lineHeight.typeSmall |
> Styling documentation generated by Cells CLI

@customElement bbva-list-bullet
@polymer
@LitElement
@demo demo/index.html
*/
export class BbvaListBullet extends LitElement {
  static get is() {
    return 'bbva-list-bullet';
  }

  static get properties() {
    return {
      /**
       * Icon to display in each item of the list
       */
      icon: {
        type: String,
      },

      /**
       * Size of the icon
       */
      iconSize: {
        type: Number,
        attribute: 'icon-size',
      },
    };
  }

  constructor() {
    super();
    this.icon = 'coronita:bullet1';
    this.iconSize = 24;
  }

  static get styles() {
    return [styles, getComponentSharedStyles('bbva-list-bullet-shared-styles')];
  }

  render() {
    return html`
      <cells-icon class="icon" size="${this.iconSize}" icon="${this.icon}"></cells-icon>
      <slot></slot>
    `;
  }
}

customElements.define(BbvaListBullet.is, BbvaListBullet);
