// @bbva-web-components/bbva-clip-card/bbva-clip-card/3.2.3/bbva-clip-card-cards.js
import { svg } from 'lit-element';

export const defaultCard = svg`
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 72 48">
  <path fill="#d3d3d3" d="M0 0h72v48H0V0z"/>
  <defs>
    <path id="a" d="M0 0h72v48H0V0z"/>
  </defs>
  <clipPath id="b">
    <use overflow="visible" xlink:href="#a"/>
  </clipPath>
  <g clip-path="url(#b)" opacity=".7">
    <radialGradient id="c" cx="-374.547" cy="1390.735" r="33.453" fx="-408" fy="1390.588" gradientTransform="matrix(.4832 .2569 .3854 -.7248 -308.985 1145.391)" gradientUnits="userSpaceOnUse">
      <stop offset=".002" stop-color="#fff"/>
      <stop offset=".985" stop-color="#e6e6e6" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#c)" d="M67.2 84L18 57.8 44.2 8.6l49.2 26.2L67.2 84z"/>
    <radialGradient id="d" cx="-405.269" cy="1390.741" r="23.962" gradientTransform="matrix(.4832 .2569 .2569 -.4832 -130.318 809.37)" gradientUnits="userSpaceOnUse">
      <stop offset=".003" stop-color="#fff"/>
      <stop offset="1" stop-color="#fff" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#d)" d="M22.5 49.7l-.1-.1 17.4-32.8.1.1-17.4 32.8z"/>
    <radialGradient id="e" cx="-3611.6" cy="-1875.684" r="33.439" fx="-3645.038" fy="-1875.831" gradientTransform="matrix(-.4832 -.2569 -.3854 .7248 -2442.038 438.51)" gradientUnits="userSpaceOnUse">
      <stop offset=".002" stop-color="#fff"/>
      <stop offset="1" stop-color="#e6e6e6" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#e)" d="M4.8-36L54-9.8 27.8 39.4l-49.2-26.2L4.8-36z"/>
    <radialGradient id="f" cx="-3642.321" cy="-1875.807" r="23.999" gradientTransform="matrix(-.4832 -.2569 -.2569 .4832 -2201.084 -14.662)" gradientUnits="userSpaceOnUse">
      <stop offset=".003" stop-color="#fff"/>
      <stop offset="1" stop-color="#fff" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#f)" d="M49.5-1.7l.1.1-17.4 32.8-.1-.1L49.5-1.7z"/>
  </g>
</svg>
`;

export const debitCard = svg`
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 72 48">
  <path d="M0 0h72v48H0V0z" fill="#02a5a5"/>
  <defs>
    <path id="a" d="M0 0h72v48H0z"/>
  </defs>
  <clipPath id="b">
    <use xlink:href="#a" overflow="visible"/>
  </clipPath>
  <g opacity=".9" clip-path="url(#b)">
    <radialGradient id="c" cx="-374.547" cy="-1342.746" r="33.455" fx="-408.002" fy="-1342.894" gradientTransform="matrix(.4832 .2569 -.3854 .7248 -290.486 1110.601)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#00cdce"/>
      <stop offset="1" stop-color="#008586" stop-opacity="0"/>
    </radialGradient>
    <path d="M67.2 84L18 57.8 44.2 8.6l49.2 26.2L67.2 84z" fill="url(#c)"/>
    <radialGradient id="d" cx="-405.269" cy="-1342.753" r="23.971" gradientTransform="matrix(.4832 .2569 -.2569 .4832 -117.986 786.177)" gradientUnits="userSpaceOnUse">
      <stop offset=".004" stop-color="#19d2d3"/>
      <stop offset="1" stop-color="#19d2d3" stop-opacity="0"/>
    </radialGradient>
    <path d="M22.5 49.7l-.1-.1 17.4-32.8.1.1-17.4 32.8z" fill="url(#d)"/>
    <radialGradient id="e" cx="-3611.6" cy="1923.673" r="33.441" fx="-3645.04" fy="1923.525" gradientTransform="matrix(-.4832 -.2569 .3854 -.7248 -2460.537 473.3)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#00cdce"/>
      <stop offset="1" stop-color="#008586" stop-opacity="0"/>
    </radialGradient>
    <path d="M4.8-36L54-9.8 27.8 39.4l-49.2-26.2L4.8-36z" fill="url(#e)"/>
    <radialGradient id="f" cx="-3642.321" cy="1923.807" r="23.999" gradientTransform="matrix(-.4832 -.2569 .2569 -.4832 -2213.415 8.531)" gradientUnits="userSpaceOnUse">
      <stop offset=".004" stop-color="#19d2d3"/>
      <stop offset="1" stop-color="#19d2d3" stop-opacity="0"/>
    </radialGradient>
    <path d="M49.5-1.7l.1.1-17.4 32.8-.1-.1L49.5-1.7z" fill="url(#f)"/>
  </g>
</svg>
`;

export const creditCard = svg`
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 72 48">
  <path fill="#043263" d="M0 0h72v48H0V0z"/>
  <defs>
    <path id="a" d="M0 0h72v48H0V0z"/>
  </defs>
  <clipPath id="b">
    <use overflow="visible" xlink:href="#a"/>
  </clipPath>
  <g clip-path="url(#b)">
    <radialGradient id="c" cx="-374.547" cy="1390.735" r="33.453" fx="-408" fy="1390.588" gradientTransform="matrix(.4832 .2569 .3854 -.7248 -308.985 1145.391)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#0064a9"/>
      <stop offset=".989" stop-color="#002e65" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#c)" d="M67.2 84L18 57.8 44.2 8.6l49.2 26.2L67.2 84z"/>
    <radialGradient id="d" cx="-405.269" cy="1390.741" r="23.962" gradientTransform="matrix(.4832 .2569 .2569 -.4832 -130.318 809.37)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#1974b1"/>
      <stop offset=".996" stop-color="#1974b1" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#d)" d="M22.5 49.7l-.1-.1 17.4-32.8.1.1-17.4 32.8z"/>
    <radialGradient id="e" cx="-3611.6" cy="-1875.684" r="33.439" fx="-3645.038" fy="-1875.831" gradientTransform="matrix(-.4832 -.2569 -.3854 .7248 -2442.038 438.51)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#0064a9"/>
      <stop offset=".989" stop-color="#002e65" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#e)" d="M4.8-36L54-9.8 27.8 39.4l-49.2-26.2L4.8-36z"/>
    <radialGradient id="f" cx="-3642.321" cy="-1875.807" r="23.999" gradientTransform="matrix(-.4832 -.2569 -.2569 .4832 -2201.084 -14.662)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#1974b1"/>
      <stop offset=".996" stop-color="#1974b1" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#f)" d="M49.5-1.7l.1.1-17.4 32.8-.1-.1L49.5-1.7z"/>
  </g>
</svg>
`;

export const prepayCard = svg`
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 72 48">
  <path fill="#1464a5" d="M0 0h72v48H0V0z"/>
  <defs>
    <path id="a" d="M0 0h72v48H0V0z"/>
  </defs>
  <clipPath id="b">
    <use overflow="visible" xlink:href="#a"/>
  </clipPath>
  <g clip-path="url(#b)" opacity=".7">
    <radialGradient id="c" cx="-374.547" cy="1390.735" r="33.453" fx="-408" fy="1390.588" gradientTransform="matrix(.4832 .2569 .3854 -.7248 -308.985 1145.391)" gradientUnits="userSpaceOnUse">
      <stop offset=".001" stop-color="#00a7e9"/>
      <stop offset=".988" stop-color="#0064a9" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#c)" d="M67.2 84L18 57.8 44.2 8.6l49.2 26.2L67.2 84z"/>
    <radialGradient id="d" cx="-405.269" cy="1390.741" r="23.962" gradientTransform="matrix(.4832 .2569 .2569 -.4832 -130.318 809.37)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#19afeb"/>
      <stop offset="1" stop-color="#19afeb" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#d)" d="M22.5 49.7l-.1-.1 17.4-32.8.1.1-17.4 32.8z"/>
    <radialGradient id="e" cx="-3611.6" cy="-1875.684" r="33.439" fx="-3645.038" fy="-1875.831" gradientTransform="matrix(-.4832 -.2569 -.3854 .7248 -2442.038 438.51)" gradientUnits="userSpaceOnUse">
      <stop offset=".001" stop-color="#00a7e9"/>
      <stop offset=".988" stop-color="#0064a9" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#e)" d="M4.8-36L54-9.8 27.8 39.4l-49.2-26.2L4.8-36z"/>
    <radialGradient id="f" cx="-3642.321" cy="-1875.807" r="23.999" gradientTransform="matrix(-.4832 -.2569 -.2569 .4832 -2201.084 -14.662)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#19afeb"/>
      <stop offset="1" stop-color="#19afeb" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#f)" d="M49.5-1.7l.1.1-17.4 32.8-.1-.1L49.5-1.7z"/>
  </g>
</svg>
`;

export const caminoCard = svg`
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 72 48">
  <path fill="#49a5e6" d="M0 0h72v48H0V0z"/>
  <defs>
    <path id="a" d="M0 0h72v48H0V0z"/>
  </defs>
  <clipPath id="b">
    <use overflow="visible" xlink:href="#a"/>
  </clipPath>
  <g clip-path="url(#b)" opacity=".45">
    <radialGradient id="c" cx="-374.547" cy="1390.735" r="33.453" fx="-408" fy="1390.588" gradientTransform="matrix(.4832 .2569 .3854 -.7248 -308.985 1145.391)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#caeefd"/>
      <stop offset=".985" stop-color="#00a7e9" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#c)" d="M67.2 84L18 57.8 44.2 8.6l49.2 26.2L67.2 84z"/>
    <radialGradient id="d" cx="-405.269" cy="1390.741" r="23.962" gradientTransform="matrix(.4832 .2569 .2569 -.4832 -130.318 809.37)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#cfeffd"/>
      <stop offset="1" stop-color="#cfeffd" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#d)" d="M22.5 49.7l-.1-.1 17.4-32.8.1.1-17.4 32.8z"/>
    <radialGradient id="e" cx="-3611.6" cy="-1875.684" r="33.439" fx="-3645.038" fy="-1875.831" gradientTransform="matrix(-.4832 -.2569 -.3854 .7248 -2442.038 438.51)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#caeefd"/>
      <stop offset="1" stop-color="#00a7e9" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#e)" d="M4.8-36L54-9.8 27.8 39.4l-49.2-26.2L4.8-36z"/>
    <radialGradient id="f" cx="-3642.321" cy="-1875.807" r="23.999" gradientTransform="matrix(-.4832 -.2569 -.2569 .4832 -2201.084 -14.662)" gradientUnits="userSpaceOnUse">
      <stop offset="0" stop-color="#cfeffd"/>
      <stop offset="1" stop-color="#cfeffd" stop-opacity="0"/>
    </radialGradient>
    <path fill="url(#f)" d="M49.5-1.7l.1.1-17.4 32.8-.1-.1L49.5-1.7z"/>
  </g>
</svg>
`;

export const goldCard = svg`
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 72 48">
  <path fill="#d0b76b" d="M0 0h72v48H0V0z"/>
  <defs>
    <path id="a" d="M0 0h72v48H0V0z"/>
  </defs>
  <clipPath id="b">
    <use overflow="visible" xlink:href="#a"/>
  </clipPath>
  <g opacity=".5">
    <g clip-path="url(#b)" opacity=".7">
      <radialGradient id="c" cx="-374.547" cy="1390.735" r="33.453" fx="-408" fy="1390.588" gradientTransform="matrix(.4832 .2569 .3854 -.7248 -308.985 1145.391)" gradientUnits="userSpaceOnUse">
        <stop offset=".002" stop-color="#fff"/>
        <stop offset=".985" stop-color="#fff" stop-opacity="0"/>
      </radialGradient>
      <path fill="url(#c)" d="M67.2 84L18 57.8 44.2 8.6l49.2 26.2L67.2 84z"/>
      <radialGradient id="d" cx="-405.269" cy="1390.741" r="23.962" gradientTransform="matrix(.4832 .2569 .2569 -.4832 -130.318 809.37)" gradientUnits="userSpaceOnUse">
        <stop offset=".003" stop-color="#fff"/>
        <stop offset="1" stop-color="#fff" stop-opacity="0"/>
      </radialGradient>
      <path fill="url(#d)" d="M22.5 49.7l-.1-.1 17.4-32.8.1.1-17.4 32.8z"/>
      <radialGradient id="e" cx="-3611.6" cy="-1875.684" r="33.439" fx="-3645.038" fy="-1875.831" gradientTransform="matrix(-.4832 -.2569 -.3854 .7248 -2442.038 438.51)" gradientUnits="userSpaceOnUse">
        <stop offset=".002" stop-color="#fff"/>
        <stop offset=".98" stop-color="#fff" stop-opacity="0"/>
      </radialGradient>
      <path fill="url(#e)" d="M4.8-36L54-9.8 27.8 39.4l-49.2-26.2L4.8-36z"/>
      <radialGradient id="f" cx="-3642.321" cy="-1875.807" r="23.999" gradientTransform="matrix(-.4832 -.2569 -.2569 .4832 -2201.084 -14.662)" gradientUnits="userSpaceOnUse">
        <stop offset=".003" stop-color="#fff"/>
        <stop offset="1" stop-color="#fff" stop-opacity="0"/>
      </radialGradient>
      <path fill="url(#f)" d="M49.5-1.7l.1.1-17.4 32.8-.1-.1L49.5-1.7z"/>
    </g>
  </g>
</svg>
`;
