// @cells-components/cells-dropdown-layer/cells-dropdown-layer-styles.js
import { css } from 'lit-element';

export default css`:host {
  display: inline-block;
  position: relative;
  box-sizing: border-box;
  --value-color: var(--cells-dropdown-layer-value-color, #565C67);
  --underline-color: var(--cells-dropdown-layer-underline-color, #C6C8CF);
  --icon-color: var(--cells-dropdown-layer-icon-color, #565C67);
  --animation-time: var(--cells-dropdown-layer-animation-time, 0.3s);
  font-family: var(--cells-fontDefault, sans-serif);
  font-size: 0.875rem;
  @apply --cells-dropdown-layer; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *::before, *::after {
  box-sizing: inherit; }

:host(.dark) {
  --value-color: var(--cells-dropdown-layer-dark-value-color, #D0D1D4);
  --underline-color: var(--cells-dropdown-layer-dark-underline-color, #565C67);
  --icon-color: var(--cells-dropdown-layer-dark-icon-color, #A3A8AD);
  @apply --cells-dropdown-layer-dark; }

.opener {
  cursor: pointer;
  @apply --cells-dropdown-layer-opener; }

.control {
  display: flex;
  align-items: center;
  position: relative;
  z-index: 1;
  @apply --cells-dropdown-layer-control; }
  .control-text {
    padding: 0 1.25rem 0.5rem 0;
    border: none;
    background-color: transparent;
    height: 1.5rem;
    line-height: 1rem;
    cursor: pointer;
    outline: none;
    text-align: left;
    transition: color var(--animation-time);
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    color: var(--value-color);
    font-size: 0.875rem;
    @apply --cells-dropdown-layer-control-text; }
  .control-prefix {
    @apply --cells-dropdown-layer-control-prefix; }
  .control-icon {
    padding: 0 0.5rem 0.5rem 0;
    height: 1.5rem;
    min-width: 1rem;
    line-height: 1rem;
    display: flex;
    align-items: center;
    justify-content: flex-start;
    color: var(--value-color);
    transition: color var(--animation-time);
    @apply --cells-dropdown-layer-control-icon; }
    .control-icon .icon {
      height: 1rem;
      width: 1rem;
      @apply --cells-dropdown-layer-control-icon-icon; }

.line {
  height: 0.5rem;
  position: relative;
  z-index: 1;
  display: flex;
  align-items: center;
  @apply --cells-dropdown-layer-line; }
  .line .underline {
    width: 100%;
    height: 0.125rem;
    overflow: hidden;
    position: relative;
    transform-origin: left center;
    transition: transform var(--animation-time);
    @apply --cells-dropdown-layer-line-underline; }
    .line .underline::after {
      content: '';
      display: block;
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      background-color: var(--underline-color);
      transform: translateX(-1rem);
      transition: transform var(--animation-time), background-color var(--animation-time);
      @apply --cells-dropdown-layer-line-underline-after; }
  .line .icons {
    position: absolute;
    bottom: 0;
    right: 0;
    top: 0;
    transition: transform var(--animation-time);
    @apply --cells-dropdown-layer-line-icons; }
    .line .icons .icon {
      width: 0.5rem;
      height: 0.5rem;
      color: var(--icon-color);
      @apply --cells-dropdown-layer-line-icon; }
      .line .icons .icon-open {
        position: absolute;
        bottom: 0;
        right: 0;
        transform-origin: top center;
        transition: transform var(--animation-time);
        @apply --cells-dropdown-layer-line-icon-open; }
      .line .icons .icon-disabled {
        border: none;
        display: none;
        @apply --cells-dropdown-layer-line-icon-disabled; }

.error-text {
  display: none;
  padding-top: 0.5rem;
  line-height: 1rem;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: var(--value-color);
  font-size: 0.75rem;
  @apply --cells-dropdown-layer-error-text; }

.layer {
  position: absolute;
  width: 100vw;
  z-index: 2;
  margin: 0 0 0.5rem;
  color: var(--cells-dropdown-layer-layer-color, #D0D1D4);
  transform: scale(0, 0);
  transition: transform var(--animation-time);
  display: none;
  @apply --cells-dropdown-layer-layer; }
  :host([animating]) .layer {
    display: flex;
    @apply --cells-dropdown-layer-layer-animating; }
  .layer-wrapper {
    max-width: var(--cells-dropdown-layer-layer-max-width, none);
    min-width: var(--cells-dropdown-layer-layer-min-width, 0);
    max-height: var(--cells-dropdown-layer-layer-max-height, none);
    min-height: var(--cells-dropdown-layer-layer-min-height, 0);
    position: relative;
    overflow: auto;
    display: flex;
    @apply --cells-dropdown-layer-wrapper; }
  .layer-content {
    background-color: var(--cells-dropdown-layer-content-bg-color, #343B49);
    padding: 1.5rem;
    max-height: 100%;
    flex: auto;
    overflow: auto;
    @apply --cells-dropdown-layer-content; }
  .layer.top {
    top: 2rem;
    @apply --cells-dropdown-layer-top; }
  .layer.bottom {
    bottom: 100%;
    @apply --cells-dropdown-layer-bottom; }
  .layer.left {
    left: 0;
    @apply --cells-dropdown-layer-left; }
    .layer.left.top {
      transform-origin: top left;
      @apply --cells-dropdown-layer-left-top; }
    .layer.left.bottom {
      transform-origin: bottom left;
      @apply --cells-dropdown-layer-left-bottom; }
  .layer.right {
    justify-content: flex-end;
    right: 0;
    @apply --cells-dropdown-layer-right; }
    .layer.right.top {
      transform-origin: top right;
      @apply --cells-dropdown-layer-right-top; }
    .layer.right.bottom {
      transform-origin: bottom right;
      @apply --cells-dropdown-layer-right-bottom; }
  .layer.centered {
    left: 50%;
    @apply --cells-dropdown-layer-centered; }
    .layer.centered .layer-wrapper {
      transform: translateX(-50%);
      @apply --cells-dropdown-layer-centered-wrapper; }
    .layer.centered.top {
      transform-origin: top left;
      @apply --cells-dropdown-layer-centered-top; }
    .layer.centered.bottom {
      transform-origin: bottom left;
      @apply --cells-dropdown-layer-centered-bottom; }

/* STATES */
:host(:hover:not([disabled]):not([opened])),
:host([focused]:not([disabled]):not([opened])) {
  --value-color: var(--cells-dropdown-layer-hovered-value-color, #224FBD);
  --underline-color: var(--cells-dropdown-layer-hovered-underline-color, #224FBD);
  --icon-color: var(--cells-dropdown-layer-hovered-icon-color, #224FBD);
  @apply --cells-dropdown-layer-hovered; }
  :host(:hover:not([disabled]):not([opened])) .line .underline:after,
  :host([focused]:not([disabled]):not([opened])) .line .underline:after {
    transform: translateX(-0.75rem);
    @apply --cells-dropdown-layer-hovered-line-underline-after; }

:host(.dark:hover:not([disabled]):not([opened])),
:host(.dark[focused]:not([disabled]):not([opened])) {
  --value-color: var(--cells-dropdown-layer-dark-hovered-value-color, #FFF);
  --underline-color: var(--cells-dropdown-layer-dark-hovered-underline-color, #FFF);
  --icon-color: var(--cells-dropdown-layer-dark-hovered-icon-color, #FFF);
  @apply --cells-dropdown-layer-dark-hovered; }

:host([opened]:not([disabled])) {
  --value-color: var(--cells-dropdown-layer-opened-value-color, #224FBD);
  --underline-color: var(--cells-dropdown-layer-opened-underline-color, #224FBD);
  --icon-color: var(--cells-dropdown-layer-opened-icon-color, #224FBD);
  @apply --cells-dropdown-layer-opened; }
  :host([opened]:not([disabled])) .line {
    @apply --cells-dropdown-layer-opened-line; }
    :host([opened]:not([disabled])) .line .underline::after {
      transform: translateX(0);
      @apply --cells-dropdown-layer-opened-line-underline-after; }
    :host([opened]:not([disabled])) .line .icons .icon-open {
      transform: rotate(180deg) translateY(0.5rem);
      @apply --cells-dropdown-layer-opened-line-icon-open; }
  :host([opened]:not([disabled])) .layer {
    z-index: 3;
    display: flex;
    transform: scale(1, 1);
    @apply --cells-dropdown-layer-opened-layer; }

:host(.dark[opened]:not([disabled])) {
  --value-color: var(--cells-dropdown-layer-dark-opened-value-color, #FFF);
  --underline-color: var(--cells-dropdown-layer-dark-opened-underline-color, #FFF);
  --icon-color: var(--cells-dropdown-layer-dark-opened-icon-color, #FFF);
  @apply --cells-dropdown-layer-dark-opened; }

:host([disabled]) {
  --value-color: var(--cells-dropdown-layer-disabled-value-color, #A3A8AD);
  --underline-color: var(--cells-dropdown-layer-disabled-underline-color,#D0D1D4);
  --icon-color: var(--cells-dropdown-layer-disabled-icon-color, #A3A8AD);
  cursor: default;
  pointer-events: none;
  @apply --cells-dropdown-layer-disabled; }
  :host([disabled]) .line .icons {
    @apply --cells-dropdown-layer-disabled-line-icons; }
    :host([disabled]) .line .icons .icon-disabled {
      display: block;
      @apply --cells-dropdown-layer-disabled-line-icon-disabled; }
    :host([disabled]) .line .icons .icon-open {
      display: none;
      @apply --cells-dropdown-layer-disabled-line-icon-open; }

:host(.dark[disabled]) {
  --value-color: var(--cells-dropdown-layer-dark-disabled-value-color, #A3A8AD);
  --underline-color: var(--cells-dropdown-layer-dark-disabled-underline-color, #565C67);
  --icon-color: var(--cells-dropdown-layer-dark-disabled-icon-color, #A3A8AD);
  @apply --cells-dropdown-layer-dark-disabled; }

:host([has-error]:not([disabled]):not([opened])) {
  --value-color: var(--cells-dropdown-layer-with-error-value-color, #F970B5);
  --underline-color: var(--cells-dropdown-layer-with-error-underline-color, #F970B5);
  --icon-color: var(--cells-dropdown-layer-with-error-icon-color, #F970B5);
  @apply --cells-dropdown-layer-with-error; }
  :host([has-error]:not([disabled]):not([opened])) .error-text {
    display: block;
    @apply --cells-dropdown-layer-with-error-text; }

:host(.dark[has-error]:not([disabled]):not([opened])) {
  --value-color: var(--cells-dropdown-layer-dark-with-error-value-color, #F970B5);
  --underline-color: var(--cells-dropdown-layer-dark-with-error-underline-color, #F970B5);
  --icon-color: var(--cells-dropdown-layer-dark-with-error-icon-color, #F970B5);
  @apply --cells-dropdown-layer-dark-with-error; }

:host(.small) {
  @apply --cells-dropdown-layer-small; }
  :host(.small) .control-text {
    height: 1rem;
    padding-bottom: 0;
    font-size: 0.75rem;
    @apply --cells-dropdown-layer-small-control-text; }
  :host(.small) .control-icon {
    height: 1rem;
    padding-bottom: 0;
    @apply --cells-dropdown-layer-small-control-icon; }
  :host(.small) .layer.top {
    top: 1.5rem;
    @apply --cells-dropdown-layer-small-layer-top; }
  :host(.small) .layer-content {
    padding: 1rem;
    @apply --cells-dropdown-layer-small-layer-content; }
  :host(.small) .error-text {
    padding-top: 0;
    line-height: 1rem;
    font-size: 0.625rem;
    @apply --cells-dropdown-layer-small-error-text; }

:host(.small[opened]:not([disabled])) {
  @apply --cells-dropdown-layer-small-opened; }
  :host(.small[opened]:not([disabled])) .line .icons .icon-open {
    transform: rotate(180deg) translateY(0.25rem);
    @apply --cells-dropdown-layer-small-opened-line-icon-open; }

:host(.icon-only) {
  @apply --cells-dropdown-layer-icon-only; }
  :host(.icon-only) .control {
    @apply --cells-dropdown-layer-icon-only-control; }
    :host(.icon-only) .control-icon {
      padding-right: 0;
      @apply --cells-dropdown-layer-icon-only-control-icon; }
    :host(.icon-only) .control-text {
      border: 0;
      clip: rect(0 0 0 0);
      height: 1px;
      margin: -1px;
      overflow: hidden;
      padding: 0;
      position: absolute;
      width: 1px;
      @apply --cells-dropdown-layer-icon-only-control-text; }
  :host(.icon-only) .line {
    @apply --cells-dropdown-layer-icon-only-line; }
    :host(.icon-only) .line .underline::after {
      transform: translateX(0);
      @apply --cells-dropdown-layer-icon-only-line-underline-after; }
    :host(.icon-only) .line .icons {
      display: none;
      @apply --cells-dropdown-layer-icon-only-line-icons; }

:host(.icon-only:hover:not([disabled]):not([opened])),
:host(.icon-only[focused]:not([disabled]):not([opened])) {
  @apply --cells-dropdown-layer-icon-only-hovered; }
  :host(.icon-only:hover:not([disabled]):not([opened])) .line .underline::after,
  :host(.icon-only[focused]:not([disabled]):not([opened])) .line .underline::after {
    transform: translateX(0);
    @apply --cells-dropdown-layer-icon-only-hovered-line-underline-after; }
`;
