// @bbva-web-components/bbva-form-select/build/test/test.js
import { c as css, u as unsafeCSS, a as colors, t as typography, l as lineHeight, f as fontFacePrimary, g as grid, b as boxShadow, C as CellsSelect, d as bbvaFoundationsStylesFocus, e as getComponentSharedStyles, h as html, i as CellsOption } from './vendor-f4325bf6.js';
export { j as axeReport, k as chaiDomDiff } from './vendor-f4325bf6.js';

var styles = css`:host {
  --_select-bg-color: var(--bbva-form-select-bg-color, var(--colorsSecondary100, ${unsafeCSS(colors.secondary100)}));
  --_select-border-color: var(--bbva-form-select-border-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_select-readonly-border-color: var(--bbva-form-select-readonly-border-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_select-button-color: var(--bbva-form-select-button-color, var(--colorsSecondary600, ${unsafeCSS(colors.secondary600)}));
  --_select-readonly-button-color: var(--bbva-form-select-readonly-button-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_select-label-color: var(--bbva-form-select-label-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_select-filled-label-color: var(--bbva-form-select-filled-label-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_select-readonly-label-color: var(--bbva-form-select-readonly-label-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_select-icon-color: var(--bbva-form-select-icon-color, var(--colorsPrimaryMedium, ${unsafeCSS(colors.primaryMedium)}));
  --_select-readonly-icon-color: var(--bbva-form-select-readonly-icon-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_message-color: var(--bbva-form-select-message-color, var(--colorsSecondary600, ${unsafeCSS(colors.secondary600)}));
  --_message-icon-color: var(--bbva-form-select-message-icon-color, var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)}));
  --_info-message-color: var(--bbva-form-select-info-message-color, var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)}));
  --_info-message-icon-color: var(--bbva-form-select-info-message-icon-color, var(--colorsPrimarySubdued, ${unsafeCSS(colors.primarySubdued)}));
  --_invalid-select-bg-color: var(--bbva-form-select-invalid-bg-color, var(--colorsTertiaryType5Lightened, ${unsafeCSS(colors.tertiaryType5Lightened)}));
  --_invalid-select-border-color: var(--bbva-form-select-invalid-border-color, var(--colorsTertiaryType5Medium, ${unsafeCSS(colors.tertiaryType5Medium)}));
  --_invalid-select-label-color: var(--bbva-form-select-invalid-label-color, var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)}));
  --_invalid-select-icon-color: var(--bbva-form-select-invalid-icon-color, var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)}));
  display: block;
  box-sizing: border-box;
  position: relative;
  @apply --bbva-form-select; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *::before, *::after {
  box-sizing: inherit; }

.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

.slot-select ::slotted(select) {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  margin: 0;
  opacity: 0;
  z-index: -1;
  pointer-events: none;
  @apply --bbva-form-select-select; }

.button {
  position: relative;
  width: 100%;
  height: 3rem;
  font-family: inherit;
  font-size: var(--typographyTypeSmall, ${unsafeCSS(typography.typeSmall)});
  line-height: var(--lineHeightTypeSmall, ${unsafeCSS(lineHeight.typeSmall)});
  font-weight: var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(fontFacePrimary.book.fontWeight)});
  background-color: var(--_select-bg-color);
  color: var(--_select-button-color);
  border: none;
  border-bottom: 1px solid var(--_select-border-color);
  border-radius: 1px;
  margin: 0;
  outline: none;
  padding: 0;
  cursor: pointer;
  text-align: left;
  @apply --bbva-form-select-button; }
  .button::-moz-focus-inner {
    border: 0; }
  .button[aria-invalid=true] {
    background-color: var(--_invalid-select-bg-color);
    border-bottom-color: var(--_invalid-select-border-color);
    @apply --bbva-form-select-invalid-button; }
  .button[disabled] {
    opacity: .4;
    cursor: default;
    @apply --bbva-form-select-disabled-button; }
  :host([readonly]) .button {
    color: var(--_select-readonly-button-color);
    border-bottom-color: var(--_select-readonly-border-color);
    cursor: default;
    @apply --bbva-form-select-readonly-button; }
  .button-content {
    display: flex;
    align-items: center;
    height: 100%;
    @apply --bbva-form-select-button-content; }
  .button-clip {
    flex: none;
    max-width: 3rem;
    margin-left: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
    display: flex;
    align-items: center;
    @apply --bbva-form-select-button-clip; }
  .button-text {
    flex: auto;
    position: relative;
    height: 100%;
    padding: calc(((2.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 15) * 1em) 0 calc(((0.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 15) * 1em);
    margin-left: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
    min-width: 0;
    @apply --bbva-form-select-button-text; }
  .button-icon {
    flex: none;
    margin-right: calc(((0.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
    width: 3rem;
    display: flex;
    align-items: center;
    justify-content: center;
    color: var(--_select-icon-color);
    @apply --bbva-form-select-button-icon; }
    [aria-invalid=true] .button-icon {
      color: var(--_invalid-select-icon-color);
      @apply --bbva-form-select-invalid-button-icon; }
    :host([readonly]) .button-icon {
      color: var(--_select-readonly-icon-color);
      @apply --bbva-form-select-readonly-button-icon; }

.label {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  color: var(--_select-label-color);
  will-change: transform;
  transform: translateY(calc(((1.5 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem));
  transform-origin: top left;
  transition: transform .2s ease-in-out;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  @apply --bbva-form-select-label; }
  [aria-invalid=true] .label {
    color: var(--_invalid-select-label-color);
    @apply --bbva-form-select-invalid-label; }
  :host([filled]) .label {
    width: 125%;
    color: var(--_select-filled-label-color);
    transform: translateY(calc(((0.3 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem)) scale(0.8);
    transition: transform .2s ease-in-out;
    @apply --bbva-form-select-filled-label; }
  :host([readonly]) .label {
    color: var(--_select-readonly-label-color);
    @apply --bbva-form-select-readonly-label; }

.content {
  display: flex;
  @apply --bbva-form-select-content; }
  .content-text {
    flex: auto;
    min-width: 0;
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    @apply --bbva-form-select-content-text; }
  .content-value {
    flex: none;
    margin-left: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
    @apply --bbva-form-select-content-value; }

.options-layer {
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 1;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 1rem;
  opacity: 0;
  visibility: hidden;
  transition: opacity .2s, visibility 0s .2s;
  @apply --bbva-form-select-options-layer; }
  .options-layer.opened {
    opacity: 1;
    visibility: visible;
    transition: opacity .2s, visibility 0s 0s;
    @apply --bbva-form-select-opened-options-layer; }

.options-overlay {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  opacity: 0.95;
  background-color: var(--colorsPrimaryCoreDarkened, ${unsafeCSS(colors.primaryCoreDarkened)});
  @apply --bbva-form-select-options-overlay; }

.options-gradient {
  position: relative;
  max-width: 18rem;
  width: 100%;
  @apply --bbva-form-select-options-gradient; }
  .options-gradient::before, .options-gradient::after {
    content: '';
    position: absolute;
    left: 0;
    right: 0;
    height: 2.5rem;
    @apply --bbva-form-select-options-gradient-element; }
  .options-gradient::before {
    top: 0;
    background: linear-gradient(to bottom, var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)}), rgba(255, 255, 255, 0));
    @apply --bbva-form-select-options-gradient-top-element; }
  .options-gradient::after {
    bottom: 0;
    background: linear-gradient(to top, var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)}), rgba(255, 255, 255, 0));
    @apply --bbva-form-select-options-gradient-bottom-element; }

.options-wrapper {
  overflow-y: auto;
  min-height: 14rem;
  max-height: 29rem;
  padding: 2.5rem 0;
  background-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  box-shadow: 0 1px 1px 0 rgba(18, 18, 18, 0.2);
  border-radius: 1px;
  @apply --bbva-form-select-options-wrapper; }

.message {
  padding: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem) calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem) 0;
  display: flex;
  align-items: center;
  font-size: var(--typographyType3XSmall, ${unsafeCSS(typography.type3XSmall)});
  line-height: var(--lineHeightType3XSmall, ${unsafeCSS(lineHeight.type3XSmall)});
  color: var(--_message-color);
  @apply --bbva-form-select-message; }
  .message.info {
    color: var(--_info-message-color);
    @apply --bbva-form-select-info-message; }
  .message-icon {
    margin-right: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
    color: var(--_message-icon-color);
    @apply --bbva-form-select-message-icon; }
    .info .message-icon {
      color: var(--_info-message-icon-color);
      @apply --bbva-form-select-info-message-icon; }

@media (min-width: 600px) {
  .options-layer {
    position: absolute;
    top: 3rem;
    bottom: auto;
    display: block;
    padding: 0;
    transition: none;
    @apply --bbva-form-select-dt-options-layer; }
    .options-layer.opened {
      transition: none;
      @apply --bbva-form-select-dt-opened-options-layer; }
  .options-overlay {
    display: none;
    @apply --bbva-form-select-dt-options-overlay; }
  .options-gradient {
    max-width: none;
    @apply --bbva-form-select-dt-options-gradient; }
    .options-gradient::before, .options-gradient::after {
      display: none;
      @apply --bbva-form-select-dt-options-gradient-element; }
  .options-wrapper {
    padding: 0;
    max-height: none;
    min-height: 9rem;
    box-shadow: var(--boxShadowType1, ${unsafeCSS(boxShadow.type1)});
    @apply --bbva-form-select-dt-options-wrapper; } }
`;

var stylesOption = css`:host {
  display: flex;
  align-items: center;
  box-sizing: border-box;
  padding: 0 calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
  font-size: var(--typographyTypeSmall, ${unsafeCSS(typography.typeSmall)});
  line-height: var(--lineHeightTypeSmall, ${unsafeCSS(lineHeight.typeSmall)});
  font-weight: var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(fontFacePrimary.book.fontWeight)});
  font-style: var(--fontFacePrimaryBookFontStyle, ${unsafeCSS(fontFacePrimary.book.fontStyle)});
  color: var(--bbva-form-option-color, var(--colorsSecondary600, ${unsafeCSS(colors.secondary600)}));
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  cursor: pointer;
  outline: none;
  @apply --bbva-form-option; }

:host([hidden]), [hidden] {
  display: none !important; }

*, *::before, *::after {
  box-sizing: inherit; }

.clip {
  flex: none;
  @apply --bbva-form-option-clip; }
  .clip ::slotted(*) {
    margin-right: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
    @apply --bbva-form-option-clip-slotted; }

.text {
  flex: auto;
  margin: 0.75rem 0;
  min-width: 0;
  white-space: normal;
  @apply --bbva-form-option-text; }

.icon {
  flex: none;
  margin-left: calc(((2 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(grid.spacer)}))) / 16) * 1rem);
  @apply --bbva-form-option-icon; }

:host([tabindex='0']) {
  position: relative;
  background-color: var(--colorsPrimaryCore, ${unsafeCSS(colors.primaryCore)});
  color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  z-index: 1;
  @apply --bbva-form-option-focused; }

:host([aria-disabled=true]) {
  opacity: .4;
  cursor: default;
  @apply --bbva-form-option-disabled; }
`;

css`[ambient^=light] {
  --bbva-form-select-bg-color: var(--colorsSecondary100, ${unsafeCSS(colors.secondary100)});
  --bbva-form-select-border-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-readonly-border-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-button-color: var(--colorsSecondary600, ${unsafeCSS(colors.secondary600)});
  --bbva-form-select-readonly-button-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-label-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-filled-label-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-readonly-label-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-icon-color: var(--colorsPrimaryMedium, ${unsafeCSS(colors.primaryMedium)});
  --bbva-form-select-readonly-icon-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-message-color: var(--colorsSecondary600, ${unsafeCSS(colors.secondary600)});
  --bbva-form-select-message-icon-color: var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)});
  --bbva-form-select-info-message-color: var(--colorsSecondary500, ${unsafeCSS(colors.secondary500)});
  --bbva-form-select-info-message-icon-color: var(--colorsPrimarySubdued, ${unsafeCSS(colors.primarySubdued)});
  --bbva-form-select-invalid-bg-color: var(--colorsTertiaryType5Lightened, ${unsafeCSS(colors.tertiaryType5Lightened)});
  --bbva-form-select-invalid-border-color: var(--colorsTertiaryType5Medium, ${unsafeCSS(colors.tertiaryType5Medium)});
  --bbva-form-select-invalid-label-color: var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)});
  --bbva-form-select-invalid-icon-color: var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)}); }

[ambient=light100] {
  --bbva-form-select-bg-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)}); }
`;

css`[ambient^=dark] {
  --bbva-form-select-bg-color: #0A5393;
  --bbva-form-select-border-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-readonly-border-color: var(--colorsSecondary100, ${unsafeCSS(colors.secondary100)});
  --bbva-form-select-button-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-readonly-button-color: var(--colorsSecondary300, ${unsafeCSS(colors.secondary300)});
  --bbva-form-select-label-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-filled-label-color: var(--colorsSecondary100, ${unsafeCSS(colors.secondary100)});
  --bbva-form-select-readonly-label-color: var(--colorsSecondary300, ${unsafeCSS(colors.secondary300)});
  --bbva-form-select-icon-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-readonly-icon-color: var(--colorsSecondary100, ${unsafeCSS(colors.secondary100)});
  --bbva-form-select-message-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-message-icon-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-info-message-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-info-message-icon-color: var(--colorsPrimaryCoreLightened, ${unsafeCSS(colors.primaryCoreLightened)});
  --bbva-form-select-invalid-bg-color: var(--colorsTertiaryType5Lightened, ${unsafeCSS(colors.tertiaryType5Lightened)});
  --bbva-form-select-invalid-border-color: var(--colorsTertiaryType5Medium, ${unsafeCSS(colors.tertiaryType5Medium)});
  --bbva-form-select-invalid-label-color: var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)});
  --bbva-form-select-invalid-icon-color: var(--colorsTertiaryType1Dark, ${unsafeCSS(colors.tertiaryType1Dark)}); }

[ambient=dark150] {
  --bbva-form-select-bg-color: var(--colorsPrimaryMedium, ${unsafeCSS(colors.primaryMedium)}); }

[ambient=dark300] {
  --bbva-form-select-bg-color: var(--colorsPrimaryCore, ${unsafeCSS(colors.primaryCore)}); }
`;

/* eslint-disable max-classes-per-file */
/**
Dropdowns are used to let users select an option from a set of choices. They should be primarily used if there are 3 or more options. For anything with fewer options consider using radio buttons or checkboxes.

`bbva-form-select` extends from `cells-select` to show an element with select dropdown functionality, providing it with BBVA Experience styles and structure. As it extends from `cells-select`, this select has all the basic select behaviors as autofocus, form participation, required, disabled & readonly attributes...

Also, it provides a custom `bbva-form-option` to be used declaratively just as a native `select` with `options`. `bbva-form-option` has a `clip` slot to provide a clip element in each option, which will be shown in the select when the option is selected.

Example:
```html
<bbva-form-select label="Default" name="select1">
  <bbva-form-option value="option1">Option 1</bbva-form-option>
  <bbva-form-option value="option2" selected>Option 2</bbva-form-option>
  <bbva-form-option value="option3">Option 3</bbva-form-option>
</bbva-form-select>
```

## Styling

The following custom properties and mixins are available for styling:

### Custom properties

| Selector                         | CSS Property                    | CSS Variable                               | Theme Variable                  | Foundations/Fallback                                      |
| -------------------------------- | ------------------------------- | ------------------------------------------ | ------------------------------- | --------------------------------------------------------- |
| .info .message-icon              | color                           | --_info-message-icon-color                 |                                 |                                                           |
| .message-icon                    | margin-right                    | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .message-icon                    | color                           | --_message-icon-color                      |                                 |                                                           |
| .message.info                    | color                           | --_info-message-color                      |                                 |                                                           |
| .message                         | padding                         | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .message                         | font-size                       |                                            | --typographyType3XSmall         | foundations.typography.type3XSmall          |
| .message                         | line-height                     |                                            | --lineHeightType3XSmall         | foundations.lineHeight.type3XSmall          |
| .message                         | color                           | --_message-color                           |                                 |                                                           |
| .options-wrapper                 | background-color                |                                            | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .options-gradient::after         | background                      |                                            | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .options-gradient::before        | background                      |                                            | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .options-overlay                 | background-color                |                                            | --colorsPrimaryCoreDarkened     | foundations.colors.primaryCoreDarkened      |
| .content-value                   | margin-left                     | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| [aria-invalid=true] .label       | color                           | --_invalid-select-label-color              |                                 |                                                           |
| :host([readonly]) .label         | color                           | --_select-readonly-label-color             |                                 |                                                           |
| :host([filled]) .label           | color                           | --_select-filled-label-color               |                                 |                                                           |
| :host([filled]) .label           | transform                       | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .label                           | color                           | --_select-label-color                      |                                 |                                                           |
| .label                           | transform                       | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| [aria-invalid=true] .button-icon | color                           | --_invalid-select-icon-color               |                                 |                                                           |
| :host([readonly]) .button-icon   | color                           | --_select-readonly-icon-color              |                                 |                                                           |
| .button-icon                     | margin-right                    | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button-icon                     | color                           | --_select-icon-color                       |                                 |                                                           |
| .button-text                     | padding                         | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button-text                     | margin-left                     | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button-clip                     | margin-left                     | --gridSpacerVariant                        | --gridSpacer                    | foundations.grid.spacer                     |
| .button[aria-invalid=true]       | background-color                | --_invalid-select-bg-color                 |                                 |                                                           |
| .button[aria-invalid=true]       | border-bottom-color             | --_invalid-select-border-color             |                                 |                                                           |
| :host([readonly]) .button        | color                           | --_select-readonly-button-color            |                                 |                                                           |
| :host([readonly]) .button        | border-bottom-color             | --_select-readonly-border-color            |                                 |                                                           |
| .button                          | font-size                       |                                            | --typographyTypeSmall           | foundations.typography.typeSmall            |
| .button                          | line-height                     |                                            | --lineHeightTypeSmall           | foundations.lineHeight.typeSmall            |
| .button                          | font-weight                     |                                            | --fontFacePrimaryBookFontWeight | foundations.fontFacePrimary.book.fontWeight |
| .button                          | background-color                | --_select-bg-color                         |                                 |                                                           |
| .button                          | color                           | --_select-button-color                     |                                 |                                                           |
| .button                          | border-bottom                   | --_select-border-color                     |                                 |                                                           |
| :host                            | --_select-bg-color              | --bbva-form-select-bg-color                | --colorsSecondary100            | foundations.colors.secondary100             |
| :host                            | --_select-border-color          | --bbva-form-select-border-color            | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-readonly-border-color | --bbva-form-select-readonly-border-color   | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-button-color          | --bbva-form-select-button-color            | --colorsSecondary600            | foundations.colors.secondary600             |
| :host                            | --_select-readonly-button-color | --bbva-form-select-readonly-button-color   | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-label-color           | --bbva-form-select-label-color             | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-filled-label-color    | --bbva-form-select-filled-label-color      | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-readonly-label-color  | --bbva-form-select-readonly-label-color    | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_select-icon-color            | --bbva-form-select-icon-color              | --colorsPrimaryMedium           | foundations.colors.primaryMedium            |
| :host                            | --_select-readonly-icon-color   | --bbva-form-select-readonly-icon-color     | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_message-color                | --bbva-form-select-message-color           | --colorsSecondary600            | foundations.colors.secondary600             |
| :host                            | --_message-icon-color           | --bbva-form-select-message-icon-color      | --colorsTertiaryType1Dark       | foundations.colors.tertiaryType1Dark        |
| :host                            | --_info-message-color           | --bbva-form-select-info-message-color      | --colorsSecondary500            | foundations.colors.secondary500             |
| :host                            | --_info-message-icon-color      | --bbva-form-select-info-message-icon-color | --colorsPrimarySubdued          | foundations.colors.primarySubdued           |
| :host                            | --_invalid-select-bg-color      | --bbva-form-select-invalid-bg-color        | --colorsTertiaryType5Lightened  | foundations.colors.tertiaryType5Lightened   |
| :host                            | --_invalid-select-border-color  | --bbva-form-select-invalid-border-color    | --colorsTertiaryType5Medium     | foundations.colors.tertiaryType5Medium      |
| :host                            | --_invalid-select-label-color   | --bbva-form-select-invalid-label-color     | --colorsTertiaryType1Dark       | foundations.colors.tertiaryType1Dark        |
| :host                            | --_invalid-select-icon-color    | --bbva-form-select-invalid-icon-color      | --colorsTertiaryType1Dark       | foundations.colors.tertiaryType1Dark        |

### @apply

| Selector                         | Mixins                                             |
| -------------------------------- | -------------------------------------------------- |
| .info .message-icon              | --bbva-form-select-info-message-icon               |
| .message-icon                    | --bbva-form-select-message-icon                    |
| .message.info                    | --bbva-form-select-info-message                    |
| .message                         | --bbva-form-select-message                         |
| .options-wrapper                 | --bbva-form-select-options-wrapper                 |
| .options-gradient::after         | --bbva-form-select-options-gradient-bottom-element |
| .options-gradient::before        | --bbva-form-select-options-gradient-top-element    |
| .options-gradient::before        | --bbva-form-select-options-gradient-element        |
| .options-gradient::after         | --bbva-form-select-options-gradient-element        |
| .options-gradient                | --bbva-form-select-options-gradient                |
| .options-overlay                 | --bbva-form-select-options-overlay                 |
| .options-layer.opened            | --bbva-form-select-opened-options-layer            |
| .options-layer                   | --bbva-form-select-options-layer                   |
| .content-value                   | --bbva-form-select-content-value                   |
| .content-text                    | --bbva-form-select-content-text                    |
| .content                         | --bbva-form-select-content                         |
| [aria-invalid=true] .label       | --bbva-form-select-invalid-label                   |
| :host([readonly]) .label         | --bbva-form-select-readonly-label                  |
| :host([filled]) .label           | --bbva-form-select-filled-label                    |
| .label                           | --bbva-form-select-label                           |
| [aria-invalid=true] .button-icon | --bbva-form-select-invalid-button-icon             |
| :host([readonly]) .button-icon   | --bbva-form-select-readonly-button-icon            |
| .button-icon                     | --bbva-form-select-button-icon                     |
| .button-text                     | --bbva-form-select-button-text                     |
| .button-clip                     | --bbva-form-select-button-clip                     |
| .button-content                  | --bbva-form-select-button-content                  |
| .button[aria-invalid=true]       | --bbva-form-select-invalid-button                  |
| .button[disabled]                | --bbva-form-select-disabled-button                 |
| :host([readonly]) .button        | --bbva-form-select-readonly-button                 |
| .button                          | --bbva-form-select-button                          |
| .slot-select ::slotted(select)   | --bbva-form-select-select                          |
| :host                            | --bbva-form-select                                 |

## Ambient

### Custom properties

| Selector          | CSS Property                               | CSS Variable | Theme Variable                 | Foundations/Fallback                                    |
| ----------------- | ------------------------------------------ | ------------ | ------------------------------ | ------------------------------------------------------- |
| [ambient=dark300] | --bbva-form-select-bg-color                |              | --colorsPrimaryCore            | foundations.colors.primaryCore            |
| [ambient=dark150] | --bbva-form-select-bg-color                |              | --colorsPrimaryMedium          | foundations.colors.primaryMedium          |
| [ambient^=dark]   | --bbva-form-select-border-color            |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-readonly-border-color   |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=dark]   | --bbva-form-select-button-color            |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-readonly-button-color   |              | --colorsSecondary300           | foundations.colors.secondary300           |
| [ambient^=dark]   | --bbva-form-select-label-color             |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-filled-label-color      |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=dark]   | --bbva-form-select-readonly-label-color    |              | --colorsSecondary300           | foundations.colors.secondary300           |
| [ambient^=dark]   | --bbva-form-select-icon-color              |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-readonly-icon-color     |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=dark]   | --bbva-form-select-message-color           |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-message-icon-color      |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-info-message-color      |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-info-message-icon-color |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=dark]   | --bbva-form-select-invalid-bg-color        |              | --colorsTertiaryType5Lightened | foundations.colors.tertiaryType5Lightened |
| [ambient^=dark]   | --bbva-form-select-invalid-border-color    |              | --colorsTertiaryType5Medium    | foundations.colors.tertiaryType5Medium    |
| [ambient^=dark]   | --bbva-form-select-invalid-label-color     |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
| [ambient^=dark]   | --bbva-form-select-invalid-icon-color      |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |

### Custom properties

| Selector           | CSS Property                               | CSS Variable | Theme Variable                 | Foundations/Fallback                                    |
| ------------------ | ------------------------------------------ | ------------ | ------------------------------ | ------------------------------------------------------- |
| [ambient=light100] | --bbva-form-select-bg-color                |              | --colorsPrimaryCoreLightened   | foundations.colors.primaryCoreLightened   |
| [ambient^=light]   | --bbva-form-select-bg-color                |              | --colorsSecondary100           | foundations.colors.secondary100           |
| [ambient^=light]   | --bbva-form-select-border-color            |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-readonly-border-color   |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-button-color            |              | --colorsSecondary600           | foundations.colors.secondary600           |
| [ambient^=light]   | --bbva-form-select-readonly-button-color   |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-label-color             |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-filled-label-color      |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-readonly-label-color    |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-icon-color              |              | --colorsPrimaryMedium          | foundations.colors.primaryMedium          |
| [ambient^=light]   | --bbva-form-select-readonly-icon-color     |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-message-color           |              | --colorsSecondary600           | foundations.colors.secondary600           |
| [ambient^=light]   | --bbva-form-select-message-icon-color      |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
| [ambient^=light]   | --bbva-form-select-info-message-color      |              | --colorsSecondary500           | foundations.colors.secondary500           |
| [ambient^=light]   | --bbva-form-select-info-message-icon-color |              | --colorsPrimarySubdued         | foundations.colors.primarySubdued         |
| [ambient^=light]   | --bbva-form-select-invalid-bg-color        |              | --colorsTertiaryType5Lightened | foundations.colors.tertiaryType5Lightened |
| [ambient^=light]   | --bbva-form-select-invalid-border-color    |              | --colorsTertiaryType5Medium    | foundations.colors.tertiaryType5Medium    |
| [ambient^=light]   | --bbva-form-select-invalid-label-color     |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
| [ambient^=light]   | --bbva-form-select-invalid-icon-color      |              | --colorsTertiaryType1Dark      | foundations.colors.tertiaryType1Dark      |
> Styling documentation generated by Cells CLI

* @customElement bbva-form-select
* @polymer
* @LitElement
* @demo demo/index.html
*/

class BbvaFormSelect extends CellsSelect {
  static get is() {
    return 'bbva-form-select';
  }
  /**
   * TagName (uppercase) of select option elements
   * @return {String} TagName of options (uppercase)
   */


  static get optionTag() {
    return 'BBVA-FORM-OPTION';
  }

  static get properties() {
    return {
      /**
       * Icon to show when select is closed
       */
      openIcon: {
        type: String,
        attribute: 'open-icon'
      },

      /**
       * Icon to show when select is open
       */
      closeIcon: {
        type: String,
        attribute: 'close-icon'
      },

      /**
       * Size for open/close icons
       */
      iconSize: {
        type: Number,
        attribute: 'icon-size'
      },

      /**
       * If true, value of selected option will be shown in select control
       */
      showValue: {
        type: Boolean,
        attribute: 'show-value'
      },

      /**
       * Icon for error message
       */
      errorIcon: {
        type: String,
        attribute: 'error-icon'
      },

      /**
       * Size for error icon
       */
      errorIconSize: {
        type: Number,
        attribute: 'error-icon-size'
      },

      /**
       * Error message
       */
      errorMessage: {
        type: String,
        attribute: 'error-message'
      },

      /**
       * Icon for info message
       */
      infoIcon: {
        type: String,
        attribute: 'info-icon'
      },

      /**
       * Size for info icon
       */
      infoIconSize: {
        type: Number,
        attribute: 'info-icon-size'
      },

      /**
       * Info message
       */
      infoMessage: {
        type: String,
        attribute: 'info-message'
      }
    };
  }

  constructor() {
    super();
    this.openIcon = 'coronita:unfold';
    this.closeIcon = 'coronita:fold';
    this.iconSize = 24;
    this.errorIcon = 'coronita:alert';
    this.errorIconSize = 16;
    this.errorMessage = '';
    this.infoIcon = 'coronita:info';
    this.infoIconSize = 16;
    this.infoMessage = '';
  }

  updated(changedProps) {
    /* istanbul ignore else */
    if (super.updated) {
      super.updated(changedProps);
    }

    if (changedProps.has('_selectedOption')) {
      if (this._selectedOptionText) {
        this.setAttribute('filled', '');
      } else {
        this.removeAttribute('filled');
      }
    }
  }

  get _errorVisible() {
    return this.errorMessage && this.invalid;
  }

  get _selectedOptionText() {
    return this._selectedOption && this._selectedOption.text;
  }

  get _selectedOptionClip() {
    return this._selectedOption && this._selectedOption.clip;
  }

  get _selectedOptionValue() {
    return this._selectedOption && this._selectedOption.value;
  }

  static get shadyStyles() {
    return `
      ${bbvaFoundationsStylesFocus('.button:focus')}
      ${styles.cssText}
      ${getComponentSharedStyles('bbva-form-select-shared-styles')}
    `;
  }

  render() {
    return html`
      <style>
        ${this.constructor.shadyStyles}
      </style>
      ${this.field} ${this.messageError} ${this.messageInfo}
    `;
  }
  /**
   * HTML for select field block
   * @return {TemplateResult} HTML of select field block
   */


  get field() {
    return html`
      ${this._control} ${this._optionsLayer} ${this._select}
    `;
  }
  /**
   * HTML for control button content
   * @return {TemplateResult} HTML of control button content
   */


  get _controlContent() {
    return html`
      <span class="button-content">
        ${this._clip}
        <span class="button-text">
          ${this._label}
          <span class="content">
            <span class="content-text">
              ${this._selectedOptionText}
            </span>
            ${this.showValue ? html`
                  <span class="content-value">
                    ${this._selectedOptionValue}
                  </span>
                ` : ''}
          </span>
        </span>
        ${this._selectIcon}
      </span>
    `;
  }
  /**
   * HTML for selected option clip, if available
   * @return {TemplateResult} HTML of selected option clip
   */


  get _clip() {
    if (this._selectedOptionClip) {
      const clip = this._selectedOptionClip.cloneNode(true);

      return html`
        <span class="button-clip" aria-hidden="true">
          ${clip}
        </span>
      `;
    }

    return '';
  }
  /**
   * HTML for select label
   * @return {TemplateResult} HTML of select label
   */


  get _label() {
    return html`
      <span id="label" class="label" aria-hidden="true"
        >${this.label}
        ${this.required ? html`
              *
            ` : ''}</span
      >
    `;
  }
  /**
   * HTML for control icon
   * @return {TemplateResult} HTML of control icon
   */


  get _selectIcon() {
    return html`
      <span class="button-icon" aria-hidden="true">
        <cells-icon
          .icon="${this.opened ? this.closeIcon : this.openIcon}"
          .size="${this.iconSize}"
        ></cells-icon>
      </span>
    `;
  }
  /**
   * HTML for options list container
   * @return {TemplateResult} HTML for options list container
   */


  get _optionsLayer() {
    return html`
      <div class="options-layer ${this.opened ? 'opened' : ''}">
        <div class="options-overlay"></div>
        <div class="options-gradient">
          <div class="options-wrapper">
            ${this._optionsList}
          </div>
        </div>
      </div>
    `;
  }
  /**
   * Template for array-provided options
   * @return {TemplateResult} HTML of items options
   */

  /* eslint-disable-next-line class-methods-use-this */


  _optionItemTemplate(item) {
    return html`
      <bbva-form-option
        ?disabled="${item.disabled}"
        ?selected="${item.selected}"
        .value="${item.value}"
        >${item.text}</bbva-form-option
      >
    `;
  }
  /**
   * Layer with error icon and error message
   * @return {TemplateResult} HTML of error message
   */


  get messageError() {
    return this._errorVisible ? html`
          <div class="message error">
            <cells-icon
              class="message-icon"
              .icon="${this.errorIcon}"
              .size="${this.errorIconSize}"
              aria-hidden="true"
            ></cells-icon>
            <span>${this.errorMessage}</span>
          </div>
        ` : '';
  }
  /**
   * Layer with info icon and info message
   * @return {TemplateResult} HTML of info message
   */


  get messageInfo() {
    return this.infoMessage ? html`
          <div class="message info">
            <cells-icon
              class="message-icon"
              .icon="${this.infoIcon}"
              .size="${this.infoIconSize}"
              aria-hidden="true"
            ></cells-icon>
            <span>${this.infoMessage}</span>
          </div>
        ` : '';
  }

}
customElements.define(BbvaFormSelect.is, BbvaFormSelect);
/*
`bbva-form-option` extends from `cells-option` to provide custom option elements aimed to be used as children of `bbva-form-select`. These options can receive text, value... as normal select options. It also provides a 'clip' slot for showing a clip element besides the option, which will be shown in the select when the option is selected.

## Styling

The following custom properties and mixins are available for styling:

### Custom properties

| Selector              | CSS Property     | CSS Variable             | Theme Variable                  | Foundations/Fallback                                      |
| --------------------- | ---------------- | ------------------------ | ------------------------------- | --------------------------------------------------------- |
| :host([tabindex='0']) | background-color |                          | --colorsPrimaryCore             | foundations.colors.primaryCore              |
| :host([tabindex='0']) | color            |                          | --colorsPrimaryCoreLightened    | foundations.colors.primaryCoreLightened     |
| .icon                 | margin-left      | --gridSpacerVariant      | --gridSpacer                    | foundations.grid.spacer                     |
| .clip ::slotted(*)    | margin-right     | --gridSpacerVariant      | --gridSpacer                    | foundations.grid.spacer                     |
| :host                 | padding          | --gridSpacerVariant      | --gridSpacer                    | foundations.grid.spacer                     |
| :host                 | font-size        |                          | --typographyTypeSmall           | foundations.typography.typeSmall            |
| :host                 | line-height      |                          | --lineHeightTypeSmall           | foundations.lineHeight.typeSmall            |
| :host                 | font-weight      |                          | --fontFacePrimaryBookFontWeight | foundations.fontFacePrimary.book.fontWeight |
| :host                 | font-style       |                          | --fontFacePrimaryBookFontStyle  | foundations.fontFacePrimary.book.fontStyle  |
| :host                 | color            | --bbva-form-option-color | --colorsSecondary600            | foundations.colors.secondary600             |

### @apply

| Selector                    | Mixins                          |
| --------------------------- | ------------------------------- |
| :host([aria-disabled=true]) | --bbva-form-option-disabled     |
| :host([tabindex='0'])       | --bbva-form-option-focused      |
| .icon                       | --bbva-form-option-icon         |
| .text                       | --bbva-form-option-text         |
| .clip ::slotted(*)          | --bbva-form-option-clip-slotted |
| .clip                       | --bbva-form-option-clip         |
| :host                       | --bbva-form-option              |

* @customElement bbva-form-option
* @polymer
* @LitElement
* @demo demo/index.html
*/

class BbvaFormOption extends CellsOption {
  static get is() {
    return 'bbva-form-option';
  }

  static get properties() {
    return {
      /**
       * Icon to show when option is selected
       */
      icon: {
        type: String
      },

      /**
       * Size for selected option icon
       */
      iconSize: {
        type: Number,
        attribute: 'icon-size'
      }
    };
  }

  constructor() {
    super();
    this.icon = 'coronita:checkmark';
    this.iconSize = 24;
  }
  /**
   * Returns node assigned in Light DOM to clip slot
   * @return { Node } Node assigned in light DOM to clip slot
   */


  get clip() {
    return this.querySelector('[slot=clip]');
  }

  static get shadyStyles() {
    return `
      ${stylesOption.cssText}
      ${getComponentSharedStyles('cells-option-shared-styles')}
    `;
  }

  render() {
    return html`
      <style>
        ${this.constructor.shadyStyles}
      </style>
      <span class="clip">
        <slot name="clip"></slot>
      </span>
      <span class="text">
        <slot></slot>
      </span>
      ${this.selected ? html`
            <span class="icon">
              <cells-icon .icon="${this.icon}" .size="${this.iconSize}"></cells-icon>
            </span>
          ` : ''}
    `;
  }

}
customElements.define(BbvaFormOption.is, BbvaFormOption);
