// @bbva-web-components/bbva-notification-help/bbva-notification-help-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  box-sizing: border-box;
  display: flex; }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

.accessible-hide {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

.icon-help {
  color: var(--bbva-notification-help-icon-color, var(--colorsPrimarySubdued, ${unsafeCSS(foundations.colors.primarySubdued)}));
  flex-shrink: 0;
  margin-right: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem); }

.text-help {
  color: var(--bbva-notification-help-text-color, var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  font-size: var(--bbva-notification-help-text-font-size, var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)}));
  line-height: var(--bbva-notification-help-text-line-height, var(--lineHeightType3XSmall, ${unsafeCSS(foundations.lineHeight.type3XSmall)}));
  margin: 0; }

:host(.alert) .icon-help {
  color: var(--bbva-notification-help-icon-color-alert, var(--colorsTertiaryType1Dark, ${unsafeCSS(foundations.colors.tertiaryType1Dark)})); }

:host(.alert) .text-help {
  color: var(--bbva-notification-help-text-color-alert, var(--colorsSecondary600, ${unsafeCSS(foundations.colors.secondary600)})); }

:host(.light) .icon-help {
  color: var(--bbva-notification-help-icon-color-light, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }

:host(.light) .text-help {
  color: var(--bbva-notification-help-text-color-light, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)})); }
`;
