// @bbva-web-components/bbva-clip-card/bbva-clip-card/bbva-clip-card-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  --_text-color: var(--bbva-clip-card-text-color,var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  --_status-color: var(--bbva-clip-card-status-color,var(--colorsTertiaryType1Dark, ${unsafeCSS(foundations.colors.tertiaryType1Dark)}));
  --_status-icon-color: var(--bbva-clip-card-status-icon-color,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_pay-compatibility-color: var(--bbva-clip-card-pay-compatibility-color,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_main-logo-color: var(--bbva-clip-card-main-logo-color,var(--colorsPrimaryCore, ${unsafeCSS(foundations.colors.primaryCore)}));
  --_main-logo-dark-color: var(--bbva-clip-card-main-logo-dark-color,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_partner-bg-color: var(--bbva-clip-card-partner-bg-color,var(--colorsSecondary100, ${unsafeCSS(foundations.colors.secondary100)}));
  --_partner-overlay-color: var(--bbva-clip-card-partner-overlay-color,var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)}));
  display: inline-block;
  position: relative;
  box-sizing: border-box;
  overflow: hidden;
  width: 4.5rem;
  height: 3rem;
  border-radius: var(--borderRadiusMedium, ${unsafeCSS(foundations.borderRadius.medium)});
  font-size: var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)});
  line-height: var(--lineHeightType3XSmall, ${unsafeCSS(foundations.lineHeight.type3XSmall)});
  font-weight: var(--fontFacePrimaryBookItalicFontWeight, ${unsafeCSS(foundations.fontFacePrimary.bookItalic.fontWeight)});
  font-style: var(--fontFacePrimaryBookItalicFontStyle, ${unsafeCSS(foundations.fontFacePrimary.bookItalic.fontStyle)});
  color: var(--_text-color); }

:host([status='blocked']) {
  --_status-color: var(--bbva-clip-card-status-blocked-color,var(--colorsTertiaryType1Dark, ${unsafeCSS(foundations.colors.tertiaryType1Dark)})); }

:host([status='off']) {
  --_status-color: var(--bbva-clip-card-status-off-color,var(--colorsSecondary500, ${unsafeCSS(foundations.colors.secondary500)})); }

:host([status='pending']) {
  --_status-color: var(--bbva-clip-card-status-pending-color,var(--colorsPrimaryCore, ${unsafeCSS(foundations.colors.primaryCore)})); }

:host([size='medium']) {
  width: 3.75rem;
  height: 2.5rem;
  font-size: var(--typographyType4XSmall, ${unsafeCSS(foundations.typography.type4XSmall)});
  line-height: var(--lineHeightType4XSmall, ${unsafeCSS(foundations.lineHeight.type4XSmall)}); }

:host([size='small']) {
  width: 3rem;
  height: 2rem;
  font-size: var(--typographyType4XSmall, ${unsafeCSS(foundations.typography.type4XSmall)});
  line-height: 0.75rem; }

:host([hidden]),
[hidden] {
  display: none !important; }

.sr-only {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px; }

*,
*:before,
*:after {
  box-sizing: inherit; }

.main {
  position: absolute;
  top: 0.3125rem;
  left: 0.25rem;
  height: 0.25rem;
  color: var(--_main-logo-color); }
  :host([size='medium']) .main {
    top: 0.25rem;
    left: 0.1875rem;
    height: 0.25rem; }
  :host([size='small']) .main {
    top: 0.1875rem;
    left: 0.125rem;
    height: 0.1875rem; }
  .main.dark {
    color: var(--_main-logo-dark-color); }

.type {
  display: flex;
  align-items: center;
  position: absolute;
  bottom: 0.25rem;
  right: 0.25rem;
  height: 0.3125rem; }
  :host([size='medium']) .type {
    bottom: 0.125rem;
    right: 0.1875rem;
    height: 0.3125rem; }
  :host([size='small']) .type {
    bottom: 0.125rem;
    right: 0.1875rem;
    height: 0.25rem; }

svg {
  height: 100%;
  fill: currentColor; }

.image-card {
  width: 100%;
  height: 100%; }

.partner {
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  height: 100%;
  padding: 0.25rem;
  background-color: var(--_partner-bg-color); }
  .partner-bg {
    width: 84.375%;
    height: 86.375%;
    position: absolute;
    bottom: 0;
    right: 0;
    background-color: var(--_partner-overlay-color);
    opacity: 0.06;
    transform: skewX(-56deg);
    transform-origin: bottom left; }
  .partner-logo {
    position: relative; }
  .partner-text {
    position: relative; }

.overlay {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 3.125rem;
  padding-left: 0.625rem; }
  :host([size='medium']) .overlay {
    width: 2.625rem;
    padding-left: 0.125rem; }
  :host([size='small']) .overlay {
    width: 2.125rem;
    padding-left: 0.125rem; }
  .overlay-bg {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: var(--_status-color);
    transform-origin: top left;
    transform: skewX(15deg); }
    .pay .overlay-bg {
      background-color: var(--_pay-compatibility-color);
      opacity: 0.9; }
    :host([size='medium']) .overlay-bg {
      transform: skewX(13deg); }
    :host([size='small']) .overlay-bg {
      transform: skewX(12deg); }
  .overlay-icon {
    position: relative;
    color: var(--_status-icon-color); }
    .pay .overlay-icon {
      height: 0.75rem; }
      :host([size='medium']) .pay .overlay-icon {
        height: 0.625rem; }
      :host([size='small']) .pay .overlay-icon {
        height: 0.625rem; }
`;
