// @bbva-web-components/bbva-foundations-styles/bbva-foundations-styles-photo-treatments.js
import { css, unsafeCSS } from 'lit-element';
import { colors } from './bbva-foundations-styles-colors.js';

/**
 * Common set of styles for Photo Treatments
 * @param  {String} selector     CSS selector for the properties
 * @param  {String} opacity      Opacity for the veil
 * @param  {String} color        Color of the veil
 * @param  {[type]} gradientTo   If defined, photo treatment will use a linear gradient to the specified direction
 * @param  {String} gradientStop Gradient stop (default 25%)
 */
export const bbvaFoundationsStylesPhotoTreatment = (
  selector = ':host',
  opacity = '30%',
  color = 'primaryCoreDarkened',
  gradientTo,
  gradientStop = '25%',
) => {
  const result = `
    ${selector} {
      content: '';
      position: absolute;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      opacity: ${opacity};
      ${
        gradientTo
          ? `
        background: linear-gradient(to ${gradientTo}, ${colors[color]} 0, transparent ${gradientStop});
      `
          : `
        background-color: ${colors[color]};
      `
      }
    }
  `;
  return css`
    ${unsafeCSS(result)}
  `;
};
