// @bbva-web-components/bbva-link/bbva-link-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  display: inline;
  box-sizing: border-box;
  border: none;
  padding: 0;
  color: var(--bbva-link-color, var(--colorsPrimaryMedium, ${unsafeCSS(foundations.colors.primaryMedium)}));
  font-weight: var(--bbva-link-font-weight, var(--fontFacePrimaryMediumFontWeight, ${unsafeCSS(foundations.fontFacePrimary.medium.fontWeight)}));
  font-style: var(--bbva-link-font-style, var(--fontFacePrimaryMediumFontStyle, ${unsafeCSS(foundations.fontFacePrimary.medium.fontStyle)}));
  font-size: var(--bbva-link-font-size, var(--typographyTypeSmall, ${unsafeCSS(foundations.typography.typeSmall)}));
  line-height: var(--bbva-link-line-height, var(--lineHeightTypeSmall, ${unsafeCSS(foundations.lineHeight.typeSmall)})); }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

:host(:focus) {
  outline: none; }

:host([disabled]) {
  opacity: var(--bbva-link-disabled-opacity, 0.4);
  color: var(--bbva-link-disabled-color, var(--colorsSecondary400, ${unsafeCSS(foundations.colors.secondary400)})); }

:host([active]),
:host([focus-visible]) {
  color: var(--bbva-link-active-color, var(--colorsPrimaryCoreDark, ${unsafeCSS(foundations.colors.primaryCoreDark)})); }

:host([icon]) {
  display: inline-flex;
  align-items: center; }

.icon {
  margin-right: calc(((1 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 16) * 1rem); }
`;
