// @bbva-web-components/bbva-clip-box/bbva-clip-box/bbva-clip-box-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles';

export default css`:host {
  display: flex;
  box-sizing: border-box;
  position: relative;
  overflow: hidden;
  width: 1em;
  height: 1em;
  border-radius: 100%;
  background: var(--bbva-clip-box-bg-color, var(--colorsPrimaryMediumLight, ${unsafeCSS(foundations.colors.primaryMediumLight)}));
  font-size: var(--bbva-clip-box-size, 3rem);
  font-weight: var(--bbva-clip-box-font-weight, var(--fontFacePrimaryBookFontWeight, ${unsafeCSS(foundations.fontFacePrimary.book.fontWeight)}));
  color: var(--bbva-clip-box-text-color, var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  transform: scale(1) translateZ(0); }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

#content {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%; }

.bitone-layer {
  width: 50%;
  height: 100%;
  position: absolute;
  top: 0;
  right: 0;
  transform-origin: 0 50%;
  transform: rotate(45deg); }
  :host(.bitone[custom]) .bitone-layer {
    background: var(--bbva-clip-box-bg-color2, var(--colorsPrimaryMediumLight, ${unsafeCSS(foundations.colors.primaryMediumLight)}));
    opacity: 1;
    mix-blend-mode: normal; }
  :host(.bitone:not([custom])) .bitone-layer {
    background: var(--bbva-clip-box-bg-color, var(--colorsPrimaryMediumLight, ${unsafeCSS(foundations.colors.primaryMediumLight)}));
    opacity: 0.3;
    mix-blend-mode: multiply; }

.initials,
.icon {
  position: relative; }

.initials {
  font-size: var(--bbva-clip-box-font-size, var(--typographyType3XSmall, ${unsafeCSS(foundations.typography.type3XSmall)})); }

:host([image]) {
  --bbva-clip-box-bg-color: transparent; }

.image {
  width: 100%;
  height: 100%; }
  .image img {
    object-fit: cover;
    object-position: center center; }

:host([size='small']) {
  --bbva-clip-box-size: 2rem; }
  :host([size='small']) .icon {
    --cells-icon-size: var(--bbva-clip-box-s-icon-size, 1rem); }

:host([size='medium']) {
  --bbva-clip-box-size: 2.5rem; }
  :host([size='medium']) .initials {
    --bbva-clip-box-font-size: var(--typographyTypeMedium, ${unsafeCSS(foundations.typography.typeMedium)}); }
  :host([size='medium']) .icon {
    --cells-icon-size: var(--bbva-clip-box-m-icon-size, 1.5rem); }

:host([size='large']) {
  --bbva-clip-box-size: 3rem; }
  :host([size='large']) .initials {
    --bbva-clip-box-font-size: var(--typographyTypeLarge, ${unsafeCSS(foundations.typography.typeLarge)}); }
  :host([size='large']) .icon {
    --cells-icon-size: var(--bbva-clip-box-l-icon-size, 1.5rem); }

:host([size='extralarge']) {
  --bbva-clip-box-size: 4rem; }
  :host([size='extralarge']) .initials {
    --bbva-clip-box-font-size: var(--typographyType2XLarge, ${unsafeCSS(foundations.typography.type2XLarge)}); }
  :host([size='extralarge']) .icon {
    --cells-icon-size: var(--bbva-clip-box-xl-icon-size, 2rem); }

:host([size='xxl']) {
  --bbva-clip-box-size: 4.5rem; }
  :host([size='xxl']) .initials {
    --bbva-clip-box-font-size: var(--typographyType3XLarge, ${unsafeCSS(foundations.typography.type3XLarge)}); }
  :host([size='xxl']) .icon {
    --cells-icon-size: var(--bbva-clip-box-xxl-icon-size, 2rem); }
`;
