// @cells-components/cells-dropdown-menu/cells-dropdown-menu-styles.js
import { css } from 'lit-element';

export default css`:host {
  display: inline-block;
  font-family: var(--font-default, sans-serif);
  box-sizing: border-box;
  --item-value-color: var(--cells-dropdown-menu-value-color, #D0D1D4);
  --item-underline-color: var(--cells-dropdown-menu-underline-color, #565C67);
  --item-highlighted-color: var(--cells-dropdown-menu-highlighted-color, #FFF);
  @apply --cells-dropdown-menu; }

*, *::before, *::after {
  font: inherit;
  box-sizing: inherit; }

:host([hidden]), [hidden] {
  display: none !important; }

:host([has-error]) {
  --cells-dropdown-layer-icon-color: var(--cells-dropdown-menu-haserror-layer-icon-color, #F970B5);
  --cells-dropdown-layer-underline-color: var(--cells-dropdown-menu-haserror-layer-underline-color, #F970B5);
  --cells-dropdown-layer-dark-icon-color: var(--cells-dropdown-menu-haserror-dark-layer-icon-color, #F970B5);
  --cells-dropdown-layer-dark-underline-color: var(--cells-dropdown-menu-haserror-dark-layer-underline-color, #F970B5);
  @apply --cells-dropdown-menu-haserror; }

.layer {
  width: 100%;
  @apply --cells-dropdown-menu-layer; }

.list,
.item {
  list-style: none;
  margin: 0;
  padding: 0;
  outline: none;
  @apply --cells-dropdown-menu-lists; }

.item {
  display: block;
  height: 2rem;
  margin-bottom: 1rem;
  cursor: pointer;
  @apply --cells-dropdown-menu-item; }
  .item.error {
    --item-value-color: var(--cells-dropdown-menu-item-error-value-color, #F970B5);
    --item-highlighted-color: var(--cells-dropdown-menu-item-error-highlighted-color, #F970B5);
    --item-underline-color: var(--cells-dropdown-menu-item-error-underline-color, #F970B5);
    @apply --cells-dropdown-menu-item-error; }
  .item:last-of-type {
    margin-bottom: 0;
    @apply --cells-dropdown-menu-item-last; }

.control {
  color: var(--item-value-color);
  height: 1rem;
  margin-bottom: 0.5rem;
  display: flex;
  padding-right: 1.5rem;
  position: relative;
  @apply --cells-dropdown-menu-control; }
  .control .icon {
    width: 0.875rem;
    height: 0.875rem;
    margin-right: 0.5rem;
    @apply --cells-dropdown-menu-control-icon; }
    .item[aria-checked="true"] .control .icon,
    .item:hover .control .icon,
    .item:focus .control .icon,
    .item.highlighted .control .icon {
      color: var(--item-highlighted-color);
      @apply --cells-dropdown-menu-control-icon-selected; }
  .control .label {
    display: inline-block;
    line-height: 1rem;
    flex: auto;
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
    @apply --cells-dropdown-menu-control-label; }
    .item[aria-checked="true"] .control .label,
    .item:hover .control .label,
    .item:focus .control .label,
    .item.highlighted .control .label {
      color: var(--item-highlighted-color);
      @apply --cells-dropdown-menu-control-label-selected; }
  .control .icon-check {
    color: var(--item-highlighted-color);
    display: none;
    width: 0.875rem;
    height: 0.875rem;
    position: absolute;
    right: 0;
    top: 0.0625rem;
    @apply --cells-dropdown-menu-control-icon-check; }
    .item[aria-checked="true"] .control .icon-check {
      display: block;
      @apply --cells-dropdown-menu-control-icon-check-checked; }

.line {
  height: 0.5rem;
  display: flex;
  align-items: center;
  position: relative;
  @apply --cells-dropdown-menu-line; }
  .line::after {
    display: block;
    content: '';
    height: 2px;
    background-color: var(--item-underline-color);
    width: 100%;
    @apply --cells-dropdown-menu-underline; }
    .item:hover .line::after,
    .item:focus .line::after,
    .item.highlighted .line::after {
      background-color: var(--item-highlighted-color);
      @apply --cells-dropdown-menu-underline-selected; }
`;
