// @bbva-web-components/bbva-button-default/bbva-button-default/bbva-button-default-styles.js
import { css, unsafeCSS } from 'lit-element';
import * as foundations from '@bbva-web-components/bbva-foundations-styles/bbva-foundations-styles.js';

export default css`:host {
  position: relative;
  width: auto;
  max-width: 100%;
  min-width: var(--bbva-button-default-min-width, 8.75rem);
  min-height: 3rem;
  padding: 0 calc(((4 * var(--gridSpacerVariant, var(--gridSpacer, ${unsafeCSS(foundations.grid.spacer)}))) / 15) * 1em);
  font-size: var(--bbva-button-default-font-size, var(--typographyTypeSmall, ${unsafeCSS(foundations.typography.typeSmall)}));
  line-height: var(--bbva-button-default-line-height, var(--lineHeightTypeSmall, ${unsafeCSS(foundations.lineHeight.typeSmall)}));
  font-weight: var(--bbva-button-default-font-weight, var(--fontFacePrimaryMediumFontWeight, ${unsafeCSS(foundations.fontFacePrimary.medium.fontWeight)}));
  text-align: center;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  border: 1px solid var(--_border-color);
  border-radius: var(--bbva-button-default-border-radius, var(--borderRadiusSmall, ${unsafeCSS(foundations.borderRadius.small)}));
  background-color: var(--_bg-color);
  color: var(--_color);
  opacity: var(--_opacity);
  transition: background-color var(--fast-transition, 0.2s), color var(--fast-transition, 0.2s); }

:host([hidden]),
[hidden] {
  display: none !important; }

*,
*:before,
*:after {
  box-sizing: inherit; }

:host(:focus) {
  outline: none; }

:host(.full-width) {
  flex-grow: 1;
  width: 100%;
  min-width: 0; }

:host([disabled]) {
  color: var(--_color); }

:host,
:host(.primary),
:host([variant='primary']) {
  --_color: var(--bbva-button-default-color,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_bg-color: var(--bbva-button-default-bg-color,var(--colorsPrimaryMedium, ${unsafeCSS(foundations.colors.primaryMedium)}));
  --_border-color: var(--bbva-button-default-border-color, transparent);
  --_opacity: var(--bbva-button-default-opacity, 1); }

:host([active]),
:host(.primary[active]),
:host([variant='primary'][active]) {
  --_color: var(--bbva-button-default-color-active,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_bg-color: var(--bbva-button-default-bg-color-active,var(--colorsPrimaryCoreLight, ${unsafeCSS(foundations.colors.primaryCoreLight)}));
  --_border-color: var(--bbva-button-default-border-color-active, transparent);
  --_opacity: var(--bbva-button-default-opacity-active, 1); }

:host([disabled]),
:host(.primary[disabled]),
:host([variant='primary'][disabled]) {
  --_color: var(--bbva-button-default-color-disabled,var(--colorsSecondary400, ${unsafeCSS(foundations.colors.secondary400)}));
  --_bg-color: var(--bbva-button-default-bg-color-disabled,var(--colorsSecondary200, ${unsafeCSS(foundations.colors.secondary200)}));
  --_border-color: var(--bbva-button-default-border-color-disabled, transparent);
  --_opacity: var(--bbva-button-default-opacity-disabled, 1); }

:host(.secondary),
:host([variant='secondary']) {
  --_color: var(--bbva-button-default-secondary-color,var(--colorsPrimaryMedium, ${unsafeCSS(foundations.colors.primaryMedium)}));
  --_bg-color: var(--bbva-button-default-secondary-bg-color, transparent);
  --_border-color: var(--bbva-button-default-secondary-border-color,var(--colorsSecondary300, ${unsafeCSS(foundations.colors.secondary300)}));
  --_opacity: var(--bbva-button-default-secondary-opacity, 1); }

:host(.secondary[active]),
:host([variant='secondary'][active]) {
  --_color: var(--bbva-button-default-secondary-color-active,var(--colorsPrimaryMedium, ${unsafeCSS(foundations.colors.primaryMedium)}));
  --_bg-color: var(--bbva-button-default-secondary-bg-color-active,var(--colorsSecondary100, ${unsafeCSS(foundations.colors.secondary100)}));
  --_border-color: var(--bbva-button-default-secondary-border-color-active, transparent);
  --_opacity: var(--bbva-button-default-secondary-opacity-active, 1); }

:host(.secondary[disabled]),
:host([variant='secondary'][disabled]) {
  --_color: var(--bbva-button-default-secondary-color-disabled,var(--colorsSecondary400, ${unsafeCSS(foundations.colors.secondary400)}));
  --_bg-color: var(--bbva-button-default-secondary-bg-color-disabled,var(--colorsSecondary200, ${unsafeCSS(foundations.colors.secondary200)}));
  --_border-color: var(--bbva-button-default-secondary-border-color-disabled, transparent);
  --_opacity: var(--bbva-button-default-secondary-opacity-disabled, 0.4); }

:host(.positive),
:host([variant='positive']) {
  --_color: var(--bbva-button-default-positive-color,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_bg-color: var(--bbva-button-default-positive-bg-color,var(--colorsPrimaryVariantDark, ${unsafeCSS(foundations.colors.primaryVariantDark)}));
  --_border-color: var(--bbva-button-default-positive-border-color, transparent);
  --_opacity: var(--bbva-button-default-positive-opacity, 1); }

:host(.positive[active]),
:host([variant='positive'][active]) {
  --_color: var(--bbva-button-default-positive-color-active,var(--colorsPrimaryCoreLightened, ${unsafeCSS(foundations.colors.primaryCoreLightened)}));
  --_bg-color: var(--bbva-button-default-positive-bg-color-active,var(--colorsPrimaryVariant, ${unsafeCSS(foundations.colors.primaryVariant)}));
  --_border-color: var(--bbva-button-default-positive-border-color-active, transparent);
  --_opacity: var(--bbva-button-default-positive-opacity-active, 1); }

:host(.positive[disabled]),
:host([variant='positive'][disabled]) {
  --_color: var(--bbva-button-default-positive-color-disabled,var(--colorsSecondary400, ${unsafeCSS(foundations.colors.secondary400)}));
  --_bg-color: var(--bbva-button-default-positive-bg-color-disabled,var(--colorsSecondary200, ${unsafeCSS(foundations.colors.secondary200)}));
  --_border-color: var(--bbva-button-default-positive-border-color-disabled, transparent);
  --_opacity: var(--bbva-button-default-positive-opacity-disabled, 0.4); }
`;
