// @cells-components/cells-amount-mixin/cells-amount-mixin.js
import { dedupingMixin } from '@cells-components/cells-lit-helpers/utils/mixin.js';
/**
 * Shared properties for amounts and currencies.
 *
 * CellsAmountMixin
 * @mixinFunction
 */
export const CellsAmountMixin = dedupingMixin(superClass => {
  return class extends superClass {
    static get properties() {
      return {
        /**
         * Amount.
         */
        amount: {
          type: Number
        },

        /**
         * ISO 4217 code for the currency
         */
        currencyCode: {
          type: String,
          attribute: 'currency-code'
        },

        /**
         * ISO 4217 for the local currency.
         * A value must be provided in order to localice the currency symbols.
         */
        localCurrency: {
          type: String,
          attribute: 'local-currency'
        },

        /**
         * Language code for the amount.
         */
        language: {
          type: String
        }
      };
    }

    constructor() {
      super();
      this.currencyCode = 'EUR';
      this.localCurrency = 'EUR';
      this.language = 'es';
    }

    /**
     * Returns if the amount is negative
     */
    get negative() {
      return this.amount < 0;
    }
  };
});
